// @swagger-api/apidom-ns-asyncapi-2/src/elements/bindings/jms/JmsServerBinding.cjs
"use strict";

exports.__esModule = true;
exports.default = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
/**
 * @public
 */
class JmsServerBinding extends _apidomCore.ObjectElement {
  constructor(content, meta, attributes) {
    super(content, meta, attributes);
    this.element = 'jmsServerBinding';
    this.classes.push('server-binding');
  }
  get jmsConnectionFactory() {
    return this.get('jmsConnectionFactory');
  }
  set jmsConnectionFactory(jmsConnectionFactory) {
    this.set('jmsConnectionFactory', jmsConnectionFactory);
  }
  get properties() {
    return this.get('properties');
  }
  set properties(properties) {
    this.set('properties', properties);
  }
  get clientID() {
    return this.get('clientID');
  }
  set clientID(clientID) {
    this.set('clientID', clientID);
  }
  get bindingVersion() {
    return this.get('bindingVersion');
  }
  set bindingVersion(bindingVersion) {
    this.set('bindingVersion', bindingVersion);
  }
}
var _default = exports.default = JmsServerBinding;
