// @swagger-api/apidom-ns-asyncapi-2/src/elements/bindings/mqtt/MqttMessageBinding.mjs
import { ObjectElement } from '@swagger-api/apidom-core';
/**
 * @public
 */
class MqttMessageBinding extends ObjectElement {
  constructor(content, meta, attributes) {
    super(content, meta, attributes);
    this.element = 'mqttMessageBinding';
    this.classes.push('message-binding');
  }
  get payloadFormatIndicator() {
    return this.get('payloadFormatIndicator');
  }
  set payloadFormatIndicator(payloadFormatIndicator) {
    this.set('payloadFormatIndicator', payloadFormatIndicator);
  }
  get correlationData() {
    return this.get('correlationData');
  }
  set correlationData(correlationData) {
    this.set('correlationData', correlationData);
  }
  get contentType() {
    return this.get('contentType');
  }
  set contentType(contentType) {
    this.set('contentType', contentType);
  }
  get responseTopic() {
    return this.get('responseTopic');
  }
  set responseTopic(responseTopic) {
    this.set('responseTopic', responseTopic);
  }
  get bindingVersion() {
    return this.get('bindingVersion');
  }
  set bindingVersion(bindingVersion) {
    this.set('bindingVersion', bindingVersion);
  }
}
export default MqttMessageBinding;
