// @swagger-api/apidom-ns-asyncapi-2/src/elements/bindings/sns/SnsChannelBinding.cjs
"use strict";

exports.__esModule = true;
exports.default = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
/**
 * @public
 */
class SnsChannelBinding extends _apidomCore.ObjectElement {
  constructor(content, meta, attributes) {
    super(content, meta, attributes);
    this.element = 'snsChannelBinding';
    this.classes.push('channel-binding');
  }
  get name() {
    return this.get('name');
  }
  set name(name) {
    this.set('name', name);
  }
  get ordering() {
    return this.get('ordering');
  }
  set ordering(ordering) {
    this.set('ordering', ordering);
  }
  get policy() {
    return this.get('policy');
  }
  set policy(policy) {
    this.set('policy', policy);
  }
  get tags() {
    return this.get('tags');
  }
  set tags(tags) {
    this.set('tags', tags);
  }
  get bindingVersion() {
    return this.get('bindingVersion');
  }
  set bindingVersion(bindingVersion) {
    this.set('bindingVersion', bindingVersion);
  }
}
var _default = exports.default = SnsChannelBinding;
