// @swagger-api/apidom-json-pointer/dist/apidom-json-pointer.browser.js
(function(global, factory) {
	typeof exports === "object" && typeof module !== "undefined" ? factory(exports) : typeof define === "function" && define.amd ? define(["exports"], factory) : (global = typeof globalThis !== "undefined" ? globalThis : global || self, factory(global.apidomJsonPointer = {}));
})(this, function(exports) {
	Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
	//#region \0rolldown/runtime.js
	var __defProp = Object.defineProperty;
	var __commonJSMin = (cb, mod) => () => (mod || (cb((mod = { exports: {} }).exports, mod), cb = null), mod.exports);
	var __exportAll = (all, no_symbols) => {
		let target = {};
		for (var name in all) __defProp(target, name, {
			get: all[name],
			enumerable: true
		});
		if (!no_symbols) __defProp(target, Symbol.toStringTag, { value: "Module" });
		return target;
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/representation/json-string.mjs
	var json_string_exports = /* @__PURE__ */ __exportAll({
		from: () => from$1,
		to: () => to$1
	});
	var to$1 = (jsonPointer) => {
		return JSON.stringify(jsonPointer);
	};
	var from$1 = (jsonString) => {
		try {
			return String(JSON.parse(jsonString));
		} catch {
			return jsonString;
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/representation/uri-fragment-identifier.mjs
	var uri_fragment_identifier_exports = /* @__PURE__ */ __exportAll({
		from: () => from,
		fromURIReference: () => fromURIReference,
		to: () => to
	});
	var to = (jsonPointer) => {
		return `#${[...jsonPointer].map((char) => /^[A-Za-z0-9\-._~!$&'()*+,;=:@/?]$/.test(char) ? char : encodeURIComponent(char)).join("")}`;
	};
	var from = (fragment) => {
		try {
			const rfc3986Fragment = fragment.startsWith("#") ? fragment.slice(1) : fragment;
			return decodeURIComponent(rfc3986Fragment);
		} catch {
			return fragment;
		}
	};
	var fromURIReference = (uriReference) => {
		const fragmentIndex = uriReference.indexOf("#");
		return from(fragmentIndex === -1 ? "#" : uriReference.substring(fragmentIndex));
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/grammar.mjs
	function grammar() {
		this.grammarObject = "grammarObject";
		this.rules = [];
		this.rules[0] = {
			name: "json-pointer",
			lower: "json-pointer",
			index: 0,
			isBkr: false
		};
		this.rules[1] = {
			name: "reference-token",
			lower: "reference-token",
			index: 1,
			isBkr: false
		};
		this.rules[2] = {
			name: "unescaped",
			lower: "unescaped",
			index: 2,
			isBkr: false
		};
		this.rules[3] = {
			name: "escaped",
			lower: "escaped",
			index: 3,
			isBkr: false
		};
		this.rules[4] = {
			name: "array-location",
			lower: "array-location",
			index: 4,
			isBkr: false
		};
		this.rules[5] = {
			name: "array-index",
			lower: "array-index",
			index: 5,
			isBkr: false
		};
		this.rules[6] = {
			name: "array-dash",
			lower: "array-dash",
			index: 6,
			isBkr: false
		};
		this.rules[7] = {
			name: "slash",
			lower: "slash",
			index: 7,
			isBkr: false
		};
		this.udts = [];
		this.rules[0].opcodes = [];
		this.rules[0].opcodes[0] = {
			type: 3,
			min: 0,
			max: Infinity
		};
		this.rules[0].opcodes[1] = {
			type: 2,
			children: [2, 3]
		};
		this.rules[0].opcodes[2] = {
			type: 4,
			index: 7
		};
		this.rules[0].opcodes[3] = {
			type: 4,
			index: 1
		};
		this.rules[1].opcodes = [];
		this.rules[1].opcodes[0] = {
			type: 3,
			min: 0,
			max: Infinity
		};
		this.rules[1].opcodes[1] = {
			type: 1,
			children: [2, 3]
		};
		this.rules[1].opcodes[2] = {
			type: 4,
			index: 2
		};
		this.rules[1].opcodes[3] = {
			type: 4,
			index: 3
		};
		this.rules[2].opcodes = [];
		this.rules[2].opcodes[0] = {
			type: 1,
			children: [
				1,
				2,
				3
			]
		};
		this.rules[2].opcodes[1] = {
			type: 5,
			min: 0,
			max: 46
		};
		this.rules[2].opcodes[2] = {
			type: 5,
			min: 48,
			max: 125
		};
		this.rules[2].opcodes[3] = {
			type: 5,
			min: 127,
			max: 1114111
		};
		this.rules[3].opcodes = [];
		this.rules[3].opcodes[0] = {
			type: 2,
			children: [1, 2]
		};
		this.rules[3].opcodes[1] = {
			type: 7,
			string: [126]
		};
		this.rules[3].opcodes[2] = {
			type: 1,
			children: [3, 4]
		};
		this.rules[3].opcodes[3] = {
			type: 7,
			string: [48]
		};
		this.rules[3].opcodes[4] = {
			type: 7,
			string: [49]
		};
		this.rules[4].opcodes = [];
		this.rules[4].opcodes[0] = {
			type: 1,
			children: [1, 2]
		};
		this.rules[4].opcodes[1] = {
			type: 4,
			index: 5
		};
		this.rules[4].opcodes[2] = {
			type: 4,
			index: 6
		};
		this.rules[5].opcodes = [];
		this.rules[5].opcodes[0] = {
			type: 1,
			children: [1, 2]
		};
		this.rules[5].opcodes[1] = {
			type: 6,
			string: [48]
		};
		this.rules[5].opcodes[2] = {
			type: 2,
			children: [3, 4]
		};
		this.rules[5].opcodes[3] = {
			type: 5,
			min: 49,
			max: 57
		};
		this.rules[5].opcodes[4] = {
			type: 3,
			min: 0,
			max: Infinity
		};
		this.rules[5].opcodes[5] = {
			type: 5,
			min: 48,
			max: 57
		};
		this.rules[6].opcodes = [];
		this.rules[6].opcodes[0] = {
			type: 7,
			string: [45]
		};
		this.rules[7].opcodes = [];
		this.rules[7].opcodes[0] = {
			type: 7,
			string: [47]
		};
		this.toString = function toString() {
			let str = "";
			str += "; JavaScript Object Notation (JSON) Pointer ABNF syntax\n";
			str += "; https://datatracker.ietf.org/doc/html/rfc6901\n";
			str += "json-pointer    = *( slash reference-token ) ; MODIFICATION: surrogate text rule used\n";
			str += "reference-token = *( unescaped / escaped )\n";
			str += "unescaped       = %x00-2E / %x30-7D / %x7F-10FFFF\n";
			str += "                ; %x2F ('/') and %x7E ('~') are excluded from 'unescaped'\n";
			str += "escaped         = \"~\" ( \"0\" / \"1\" )\n";
			str += "                ; representing '~' and '/', respectively\n";
			str += "\n";
			str += "; https://datatracker.ietf.org/doc/html/rfc6901#section-4\n";
			str += "array-location  = array-index / array-dash\n";
			str += "array-index     = %x30 / ( %x31-39 *(%x30-39) )\n";
			str += "                ; \"0\", or digits without a leading \"0\"\n";
			str += "array-dash      = \"-\"\n";
			str += "\n";
			str += "; Surrogate named rules\n";
			str += "slash           = \"/\"\n";
			return str;
		};
	}
	//#endregion
	//#region ../../node_modules/apg-lite/lib/parser.js
	var Parser = function fnparser() {
		const id = identifiers;
		const utils = utilities;
		const p = this;
		const thisFileName = "parser.js: Parser(): ";
		const systemData = function systemData() {
			this.state = id.ACTIVE;
			this.phraseLength = 0;
			this.refresh = () => {
				this.state = id.ACTIVE;
				this.phraseLength = 0;
			};
		};
		p.ast = void 0;
		p.stats = void 0;
		p.trace = void 0;
		p.callbacks = [];
		let lookAhead = 0;
		let treeDepth = 0;
		let maxTreeDepth = 0;
		let nodeHits = 0;
		let maxMatched = 0;
		let rules = void 0;
		let udts = void 0;
		let opcodes = void 0;
		let chars = void 0;
		let sysData = new systemData();
		let ruleCallbacks = void 0;
		let udtCallbacks = void 0;
		let userData = void 0;
		const clear = () => {
			lookAhead = 0;
			treeDepth = 0;
			maxTreeDepth = 0;
			nodeHits = 0;
			maxMatched = 0;
			rules = void 0;
			udts = void 0;
			opcodes = void 0;
			chars = void 0;
			sysData.refresh();
			ruleCallbacks = void 0;
			udtCallbacks = void 0;
			userData = void 0;
		};
		const initializeCallbacks = () => {
			const functionName = `${thisFileName}initializeCallbacks(): `;
			let i;
			ruleCallbacks = [];
			udtCallbacks = [];
			for (i = 0; i < rules.length; i += 1) ruleCallbacks[i] = void 0;
			for (i = 0; i < udts.length; i += 1) udtCallbacks[i] = void 0;
			let func;
			const list = [];
			for (i = 0; i < rules.length; i += 1) list.push(rules[i].lower);
			for (i = 0; i < udts.length; i += 1) list.push(udts[i].lower);
			for (const index in p.callbacks) if (p.callbacks.hasOwnProperty(index)) {
				i = list.indexOf(index.toLowerCase());
				if (i < 0) throw new Error(`${functionName}syntax callback '${index}' not a rule or udt name`);
				func = p.callbacks[index] ? p.callbacks[index] : void 0;
				if (typeof func === "function" || func === void 0) if (i < rules.length) ruleCallbacks[i] = func;
				else udtCallbacks[i - rules.length] = func;
				else throw new Error(`${functionName}syntax callback[${index}] must be function reference or falsy)`);
			}
		};
		p.parse = (grammar, startName, inputString, callbackData) => {
			const functionName = `${thisFileName}parse(): `;
			clear();
			chars = utils.stringToChars(inputString);
			rules = grammar.rules;
			udts = grammar.udts;
			const lower = startName.toLowerCase();
			let startIndex = void 0;
			for (const i in rules) if (rules.hasOwnProperty(i)) {
				if (lower === rules[i].lower) {
					startIndex = rules[i].index;
					break;
				}
			}
			if (startIndex === void 0) throw new Error(`${functionName}start rule name '${startRule}' not recognized`);
			initializeCallbacks();
			if (p.trace) p.trace.init(rules, udts, chars);
			if (p.stats) p.stats.init(rules, udts);
			if (p.ast) p.ast.init(rules, udts, chars);
			userData = callbackData;
			opcodes = [{
				type: id.RNM,
				index: startIndex
			}];
			opExecute(0, 0);
			opcodes = void 0;
			let success = false;
			switch (sysData.state) {
				case id.ACTIVE: throw new Error(`${functionName}final state should never be 'ACTIVE'`);
				case id.NOMATCH:
					success = false;
					break;
				case id.EMPTY:
				case id.MATCH:
					if (sysData.phraseLength === chars.length) success = true;
					else success = false;
					break;
				default: throw new Error("unrecognized state");
			}
			return {
				success,
				state: sysData.state,
				stateName: id.idName(sysData.state),
				length: chars.length,
				matched: sysData.phraseLength,
				maxMatched,
				maxTreeDepth,
				nodeHits
			};
		};
		const opALT = (opIndex, phraseIndex) => {
			const op = opcodes[opIndex];
			for (let i = 0; i < op.children.length; i += 1) {
				opExecute(op.children[i], phraseIndex);
				if (sysData.state !== id.NOMATCH) break;
			}
		};
		const opCAT = (opIndex, phraseIndex) => {
			let success;
			let astLength;
			let catCharIndex;
			let catPhrase;
			const op = opcodes[opIndex];
			if (p.ast) astLength = p.ast.getLength();
			success = true;
			catCharIndex = phraseIndex;
			catPhrase = 0;
			for (let i = 0; i < op.children.length; i += 1) {
				opExecute(op.children[i], catCharIndex);
				if (sysData.state === id.NOMATCH) {
					success = false;
					break;
				} else {
					catCharIndex += sysData.phraseLength;
					catPhrase += sysData.phraseLength;
				}
			}
			if (success) {
				sysData.state = catPhrase === 0 ? id.EMPTY : id.MATCH;
				sysData.phraseLength = catPhrase;
			} else {
				sysData.state = id.NOMATCH;
				sysData.phraseLength = 0;
				if (p.ast) p.ast.setLength(astLength);
			}
		};
		const opREP = (opIndex, phraseIndex) => {
			let astLength;
			let repCharIndex;
			let repPhrase;
			let repCount;
			const op = opcodes[opIndex];
			if (op.max === 0) {
				sysData.state = id.EMPTY;
				sysData.phraseLength = 0;
				return;
			}
			repCharIndex = phraseIndex;
			repPhrase = 0;
			repCount = 0;
			if (p.ast) astLength = p.ast.getLength();
			while (1) {
				if (repCharIndex >= chars.length) break;
				opExecute(opIndex + 1, repCharIndex);
				if (sysData.state === id.NOMATCH) break;
				if (sysData.state === id.EMPTY) break;
				repCount += 1;
				repPhrase += sysData.phraseLength;
				repCharIndex += sysData.phraseLength;
				if (repCount === op.max) break;
			}
			if (sysData.state === id.EMPTY) {
				sysData.state = repPhrase === 0 ? id.EMPTY : id.MATCH;
				sysData.phraseLength = repPhrase;
			} else if (repCount >= op.min) {
				sysData.state = repPhrase === 0 ? id.EMPTY : id.MATCH;
				sysData.phraseLength = repPhrase;
			} else {
				sysData.state = id.NOMATCH;
				sysData.phraseLength = 0;
				if (p.ast) p.ast.setLength(astLength);
			}
		};
		const validateRnmCallbackResult = (rule, sysData, charsLeft, down) => {
			if (sysData.phraseLength > charsLeft) {
				let str = `${thisFileName}opRNM(${rule.name}): callback function error: `;
				str += `sysData.phraseLength: ${sysData.phraseLength}`;
				str += ` must be <= remaining chars: ${charsLeft}`;
				throw new Error(str);
			}
			switch (sysData.state) {
				case id.ACTIVE:
					if (!down) throw new Error(`${thisFileName}opRNM(${rule.name}): callback function return error. ACTIVE state not allowed.`);
					break;
				case id.EMPTY:
					sysData.phraseLength = 0;
					break;
				case id.MATCH:
					if (sysData.phraseLength === 0) sysData.state = id.EMPTY;
					break;
				case id.NOMATCH:
					sysData.phraseLength = 0;
					break;
				default: throw new Error(`${thisFileName}opRNM(${rule.name}): callback function return error. Unrecognized return state: ${sysData.state}`);
			}
		};
		const opRNM = (opIndex, phraseIndex) => {
			let astLength;
			let astDefined;
			let savedOpcodes;
			const op = opcodes[opIndex];
			const rule = rules[op.index];
			const callback = ruleCallbacks[rule.index];
			if (!lookAhead) {
				astDefined = p.ast && p.ast.ruleDefined(op.index);
				if (astDefined) {
					astLength = p.ast.getLength();
					p.ast.down(op.index, rules[op.index].name);
				}
			}
			if (callback) {
				const charsLeft = chars.length - phraseIndex;
				callback(sysData, chars, phraseIndex, userData);
				validateRnmCallbackResult(rule, sysData, charsLeft, true);
				if (sysData.state === id.ACTIVE) {
					savedOpcodes = opcodes;
					opcodes = rule.opcodes;
					opExecute(0, phraseIndex);
					opcodes = savedOpcodes;
					callback(sysData, chars, phraseIndex, userData);
					validateRnmCallbackResult(rule, sysData, charsLeft, false);
				}
			} else {
				savedOpcodes = opcodes;
				opcodes = rule.opcodes;
				opExecute(0, phraseIndex, sysData);
				opcodes = savedOpcodes;
			}
			if (!lookAhead) {
				if (astDefined) if (sysData.state === id.NOMATCH) p.ast.setLength(astLength);
				else p.ast.up(op.index, rule.name, phraseIndex, sysData.phraseLength);
			}
		};
		const opTRG = (opIndex, phraseIndex) => {
			const op = opcodes[opIndex];
			sysData.state = id.NOMATCH;
			if (phraseIndex < chars.length) {
				if (op.min <= chars[phraseIndex] && chars[phraseIndex] <= op.max) {
					sysData.state = id.MATCH;
					sysData.phraseLength = 1;
				}
			}
		};
		const opTBS = (opIndex, phraseIndex) => {
			const op = opcodes[opIndex];
			const len = op.string.length;
			sysData.state = id.NOMATCH;
			if (phraseIndex + len <= chars.length) {
				for (let i = 0; i < len; i += 1) if (chars[phraseIndex + i] !== op.string[i]) return;
				sysData.state = id.MATCH;
				sysData.phraseLength = len;
			}
		};
		const opTLS = (opIndex, phraseIndex) => {
			let code;
			const op = opcodes[opIndex];
			sysData.state = id.NOMATCH;
			const len = op.string.length;
			if (len === 0) {
				sysData.state = id.EMPTY;
				return;
			}
			if (phraseIndex + len <= chars.length) {
				for (let i = 0; i < len; i += 1) {
					code = chars[phraseIndex + i];
					if (code >= 65 && code <= 90) code += 32;
					if (code !== op.string[i]) return;
				}
				sysData.state = id.MATCH;
				sysData.phraseLength = len;
			}
		};
		const validateUdtCallbackResult = (udt, sysData, charsLeft) => {
			if (sysData.phraseLength > charsLeft) {
				let str = `${thisFileName}opUDT(${udt.name}): callback function error: `;
				str += `sysData.phraseLength: ${sysData.phraseLength}`;
				str += ` must be <= remaining chars: ${charsLeft}`;
				throw new Error(str);
			}
			switch (sysData.state) {
				case id.ACTIVE: throw new Error(`${thisFileName}opUDT(${udt.name}) ACTIVE state return not allowed.`);
				case id.EMPTY:
					if (udt.empty) sysData.phraseLength = 0;
					else throw new Error(`${thisFileName}opUDT(${udt.name}) may not return EMPTY.`);
					break;
				case id.MATCH:
					if (sysData.phraseLength === 0) if (udt.empty) sysData.state = id.EMPTY;
					else throw new Error(`${thisFileName}opUDT(${udt.name}) may not return EMPTY.`);
					break;
				case id.NOMATCH:
					sysData.phraseLength = 0;
					break;
				default: throw new Error(`${thisFileName}opUDT(${udt.name}): callback function return error. Unrecognized return state: ${sysData.state}`);
			}
		};
		const opUDT = (opIndex, phraseIndex) => {
			let astLength;
			let astIndex;
			let astDefined;
			const op = opcodes[opIndex];
			const udt = udts[op.index];
			sysData.UdtIndex = udt.index;
			if (!lookAhead) {
				astDefined = p.ast && p.ast.udtDefined(op.index);
				if (astDefined) {
					astIndex = rules.length + op.index;
					astLength = p.ast.getLength();
					p.ast.down(astIndex, udt.name);
				}
			}
			const charsLeft = chars.length - phraseIndex;
			udtCallbacks[op.index](sysData, chars, phraseIndex, userData);
			validateUdtCallbackResult(udt, sysData, charsLeft);
			if (!lookAhead) {
				if (astDefined) if (sysData.state === id.NOMATCH) p.ast.setLength(astLength);
				else p.ast.up(astIndex, udt.name, phraseIndex, sysData.phraseLength);
			}
		};
		const opAND = (opIndex, phraseIndex) => {
			lookAhead += 1;
			opExecute(opIndex + 1, phraseIndex);
			lookAhead -= 1;
			sysData.phraseLength = 0;
			switch (sysData.state) {
				case id.EMPTY:
					sysData.state = id.EMPTY;
					break;
				case id.MATCH:
					sysData.state = id.EMPTY;
					break;
				case id.NOMATCH:
					sysData.state = id.NOMATCH;
					break;
				default: throw new Error(`opAND: invalid state ${sysData.state}`);
			}
		};
		const opNOT = (opIndex, phraseIndex) => {
			lookAhead += 1;
			opExecute(opIndex + 1, phraseIndex);
			lookAhead -= 1;
			sysData.phraseLength = 0;
			switch (sysData.state) {
				case id.EMPTY:
				case id.MATCH:
					sysData.state = id.NOMATCH;
					break;
				case id.NOMATCH:
					sysData.state = id.EMPTY;
					break;
				default: throw new Error(`opNOT: invalid state ${sysData.state}`);
			}
		};
		const opExecute = (opIndex, phraseIndex) => {
			const functionName = `${thisFileName}opExecute(): `;
			const op = opcodes[opIndex];
			nodeHits += 1;
			if (treeDepth > maxTreeDepth) maxTreeDepth = treeDepth;
			treeDepth += 1;
			sysData.refresh();
			if (p.trace) p.trace.down(op, phraseIndex);
			switch (op.type) {
				case id.ALT:
					opALT(opIndex, phraseIndex);
					break;
				case id.CAT:
					opCAT(opIndex, phraseIndex);
					break;
				case id.REP:
					opREP(opIndex, phraseIndex);
					break;
				case id.RNM:
					opRNM(opIndex, phraseIndex);
					break;
				case id.TRG:
					opTRG(opIndex, phraseIndex);
					break;
				case id.TBS:
					opTBS(opIndex, phraseIndex);
					break;
				case id.TLS:
					opTLS(opIndex, phraseIndex);
					break;
				case id.UDT:
					opUDT(opIndex, phraseIndex);
					break;
				case id.AND:
					opAND(opIndex, phraseIndex);
					break;
				case id.NOT:
					opNOT(opIndex, phraseIndex);
					break;
				default: throw new Error(`${functionName}unrecognized operator`);
			}
			if (!lookAhead) {
				if (phraseIndex + sysData.phraseLength > maxMatched) maxMatched = phraseIndex + sysData.phraseLength;
			}
			if (p.stats) p.stats.collect(op, sysData);
			if (p.trace) p.trace.up(op, sysData.state, phraseIndex, sysData.phraseLength);
			treeDepth -= 1;
		};
	};
	var Ast = function fnast() {
		const thisFileName = "parser.js: Ast()): ";
		const id = identifiers;
		const utils = utilities;
		const a = this;
		let rules = void 0;
		let udts = void 0;
		let chars = void 0;
		let nodeCount = 0;
		const nodeCallbacks = [];
		const stack = [];
		const records = [];
		a.callbacks = [];
		a.init = (rulesIn, udtsIn, charsIn) => {
			stack.length = 0;
			records.length = 0;
			nodeCount = 0;
			rules = rulesIn;
			udts = udtsIn;
			chars = charsIn;
			let i;
			const list = [];
			for (i = 0; i < rules.length; i += 1) list.push(rules[i].lower);
			for (i = 0; i < udts.length; i += 1) list.push(udts[i].lower);
			nodeCount = rules.length + udts.length;
			for (i = 0; i < nodeCount; i += 1) nodeCallbacks[i] = void 0;
			for (const index in a.callbacks) if (a.callbacks.hasOwnProperty(index)) {
				const lower = index.toLowerCase();
				i = list.indexOf(lower);
				if (i < 0) throw new Error(`${thisFileName}init: node '${index}' not a rule or udt name`);
				nodeCallbacks[i] = a.callbacks[index];
			}
		};
		a.ruleDefined = (index) => !!nodeCallbacks[index];
		a.udtDefined = (index) => !!nodeCallbacks[rules.length + index];
		a.down = (callbackIndex, name) => {
			const thisIndex = records.length;
			stack.push(thisIndex);
			records.push({
				name,
				thisIndex,
				thatIndex: void 0,
				state: id.SEM_PRE,
				callbackIndex,
				phraseIndex: void 0,
				phraseLength: void 0,
				stack: stack.length
			});
			return thisIndex;
		};
		a.up = (callbackIndex, name, phraseIndex, phraseLength) => {
			const thisIndex = records.length;
			const thatIndex = stack.pop();
			records.push({
				name,
				thisIndex,
				thatIndex,
				state: id.SEM_POST,
				callbackIndex,
				phraseIndex,
				phraseLength,
				stack: stack.length
			});
			records[thatIndex].thatIndex = thisIndex;
			records[thatIndex].phraseIndex = phraseIndex;
			records[thatIndex].phraseLength = phraseLength;
			return thisIndex;
		};
		a.translate = (data) => {
			let callback;
			let record;
			for (let i = 0; i < records.length; i += 1) {
				record = records[i];
				callback = nodeCallbacks[record.callbackIndex];
				if (callback) {
					if (record.state === id.SEM_PRE) callback(id.SEM_PRE, chars, record.phraseIndex, record.phraseLength, data);
					else if (callback) callback(id.SEM_POST, chars, record.phraseIndex, record.phraseLength, data);
				}
			}
		};
		a.setLength = (length) => {
			records.length = length;
			if (length > 0) stack.length = records[length - 1].stack;
			else stack.length = 0;
		};
		a.getLength = () => records.length;
		function indent(n) {
			let ret = "";
			while (n-- > 0) ret += " ";
			return ret;
		}
		a.toXml = () => {
			let xml = "";
			let depth = 0;
			xml += "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n";
			xml += `<root nodes="${records.length / 2}" characters="${chars.length}">\n`;
			xml += `<!-- input string -->\n`;
			xml += indent(depth + 2);
			xml += utils.charsToString(chars);
			xml += "\n";
			records.forEach((rec) => {
				if (rec.state === id.SEM_PRE) {
					depth += 1;
					xml += indent(depth);
					xml += `<node name="${rec.name}" index="${rec.phraseIndex}" length="${rec.phraseLength}">\n`;
					xml += indent(depth + 2);
					xml += utils.charsToString(chars, rec.phraseIndex, rec.phraseLength);
					xml += "\n";
				} else {
					xml += indent(depth);
					xml += `</node><!-- name="${rec.name}" -->\n`;
					depth -= 1;
				}
			});
			xml += "</root>\n";
			return xml;
		};
	};
	var Trace$1 = function fntrace() {
		const id = identifiers;
		const utils = utilities;
		const thisFile = "parser.js: Trace(): ";
		let chars = void 0;
		let rules = void 0;
		let udts = void 0;
		let out = "";
		let treeDepth = 0;
		const MAX_PHRASE = 100;
		const t = this;
		const indent = (n) => {
			let ret = "";
			let count = 0;
			if (n >= 0) while (n--) {
				count += 1;
				if (count === 5) {
					ret += "|";
					count = 0;
				} else ret += ".";
			}
			return ret;
		};
		t.init = (r, u, c) => {
			rules = r;
			udts = u;
			chars = c;
		};
		const opName = (op) => {
			let name;
			switch (op.type) {
				case id.ALT:
					name = "ALT";
					break;
				case id.CAT:
					name = "CAT";
					break;
				case id.REP:
					if (op.max === Infinity) name = `REP(${op.min},inf)`;
					else name = `REP(${op.min},${op.max})`;
					break;
				case id.RNM:
					name = `RNM(${rules[op.index].name})`;
					break;
				case id.TRG:
					name = `TRG(${op.min},${op.max})`;
					break;
				case id.TBS:
					if (op.string.length > 6) name = `TBS(${utils.charsToString(op.string, 0, 3)}...)`;
					else name = `TBS(${utils.charsToString(op.string, 0, 6)})`;
					break;
				case id.TLS:
					if (op.string.length > 6) name = `TLS(${utils.charsToString(op.string, 0, 3)}...)`;
					else name = `TLS(${utils.charsToString(op.string, 0, 6)})`;
					break;
				case id.UDT:
					name = `UDT(${udts[op.index].name})`;
					break;
				case id.AND:
					name = "AND";
					break;
				case id.NOT:
					name = "NOT";
					break;
				default: throw new Error(`${thisFile}Trace: opName: unrecognized opcode`);
			}
			return name;
		};
		t.down = (op, offset) => {
			const lead = indent(treeDepth);
			const len = Math.min(MAX_PHRASE, chars.length - offset);
			let phrase = utils.charsToString(chars, offset, len);
			if (len < chars.length - offset) phrase += "...";
			phrase = `${lead}|-|[${opName(op)}]${phrase}\n`;
			out += phrase;
			treeDepth += 1;
		};
		t.up = (op, state, offset, phraseLength) => {
			const thisFunc = `${thisFile}trace.up: `;
			treeDepth -= 1;
			const lead = indent(treeDepth);
			let len;
			let phrase;
			let st;
			switch (state) {
				case id.EMPTY:
					st = "|E|";
					phrase = `''`;
					break;
				case id.MATCH:
					st = "|M|";
					len = Math.min(MAX_PHRASE, phraseLength);
					if (len < phraseLength) phrase = `'${utils.charsToString(chars, offset, len)}...'`;
					else phrase = `'${utils.charsToString(chars, offset, len)}'`;
					break;
				case id.NOMATCH:
					st = "|N|";
					phrase = "";
					break;
				default: throw new Error(`${thisFunc} unrecognized state`);
			}
			phrase = `${lead}${st}[${opName(op)}]${phrase}\n`;
			out += phrase;
		};
		t.displayTrace = () => out;
	};
	var Stats = function fnstats() {
		const id = identifiers;
		const thisFileName = "parser.js: Stats(): ";
		let rules;
		let udts;
		let totals;
		const stats = [];
		const ruleStats = [];
		const udtStats = [];
		this.init = (r, u) => {
			rules = r;
			udts = u;
			clear();
		};
		this.collect = (op, sys) => {
			incStat(totals, sys.state, sys.phraseLength);
			incStat(stats[op.type], sys.state, sys.phraseLength);
			if (op.type === id.RNM) incStat(ruleStats[op.index], sys.state, sys.phraseLength);
			if (op.type === id.UDT) incStat(udtStats[op.index], sys.state, sys.phraseLength);
		};
		this.displayStats = () => {
			let out = "";
			const totals = {
				match: 0,
				empty: 0,
				nomatch: 0,
				total: 0
			};
			const displayRow = (op, m, e, n, t) => {
				totals.match += m;
				totals.empty += e;
				totals.nomatch += n;
				totals.total += t;
				return `${op} | ${normalize(m)} | ${normalize(e)} | ${normalize(n)} | ${normalize(t)} |\n`;
			};
			out += "          OPERATOR STATS\n";
			out += "      |   MATCH |   EMPTY | NOMATCH |   TOTAL |\n";
			out += displayRow("  ALT", stats[id.ALT].match, stats[id.ALT].empty, stats[id.ALT].nomatch, stats[id.ALT].total);
			out += displayRow("  CAT", stats[id.CAT].match, stats[id.CAT].empty, stats[id.CAT].nomatch, stats[id.CAT].total);
			out += displayRow("  REP", stats[id.REP].match, stats[id.REP].empty, stats[id.REP].nomatch, stats[id.REP].total);
			out += displayRow("  RNM", stats[id.RNM].match, stats[id.RNM].empty, stats[id.RNM].nomatch, stats[id.RNM].total);
			out += displayRow("  TRG", stats[id.TRG].match, stats[id.TRG].empty, stats[id.TRG].nomatch, stats[id.TRG].total);
			out += displayRow("  TBS", stats[id.TBS].match, stats[id.TBS].empty, stats[id.TBS].nomatch, stats[id.TBS].total);
			out += displayRow("  TLS", stats[id.TLS].match, stats[id.TLS].empty, stats[id.TLS].nomatch, stats[id.TLS].total);
			out += displayRow("  UDT", stats[id.UDT].match, stats[id.UDT].empty, stats[id.UDT].nomatch, stats[id.UDT].total);
			out += displayRow("  AND", stats[id.AND].match, stats[id.AND].empty, stats[id.AND].nomatch, stats[id.AND].total);
			out += displayRow("  NOT", stats[id.NOT].match, stats[id.NOT].empty, stats[id.NOT].nomatch, stats[id.NOT].total);
			out += displayRow("TOTAL", totals.match, totals.empty, totals.nomatch, totals.total);
			return out;
		};
		this.displayHits = (type) => {
			let out = "";
			const displayRow = (m, e, n, t, name) => {
				totals.match += m;
				totals.empty += e;
				totals.nomatch += n;
				totals.total += t;
				return `| ${normalize(m)} | ${normalize(e)} | ${normalize(n)} | ${normalize(t)} | ${name}\n`;
			};
			if (typeof type === "string" && type.toLowerCase()[0] === "a") {
				ruleStats.sort(sortAlpha);
				udtStats.sort(sortAlpha);
				out += "    RULES/UDTS ALPHABETICALLY\n";
			} else if (typeof type === "string" && type.toLowerCase()[0] === "i") {
				ruleStats.sort(sortIndex);
				udtStats.sort(sortIndex);
				out += "    RULES/UDTS BY INDEX\n";
			} else {
				ruleStats.sort(sortHits);
				udtStats.sort(sortHits);
				out += "    RULES/UDTS BY HIT COUNT\n";
			}
			out += "|   MATCH |   EMPTY | NOMATCH |   TOTAL | NAME\n";
			for (let i = 0; i < ruleStats.length; i += 1) {
				let r = ruleStats[i];
				if (r.total) out += displayRow(r.match, r.empty, r.nomatch, r.total, r.name);
			}
			for (let i = 0; i < udtStats.length; i += 1) {
				let r = udtStats[i];
				if (r.total) out += displayRow(r.match, r.empty, r.nomatch, r.total, r.name);
			}
			return out;
		};
		const normalize = (n) => {
			if (n < 10) return `      ${n}`;
			if (n < 100) return `     ${n}`;
			if (n < 1e3) return `    ${n}`;
			if (n < 1e4) return `   ${n}`;
			if (n < 1e5) return `  ${n}`;
			if (n < 1e6) return ` ${n}`;
			return `${n}`;
		};
		const sortAlpha = (lhs, rhs) => {
			if (lhs.lower < rhs.lower) return -1;
			if (lhs.lower > rhs.lower) return 1;
			return 0;
		};
		const sortHits = (lhs, rhs) => {
			if (lhs.total < rhs.total) return 1;
			if (lhs.total > rhs.total) return -1;
			return sortAlpha(lhs, rhs);
		};
		const sortIndex = (lhs, rhs) => {
			if (lhs.index < rhs.index) return -1;
			if (lhs.index > rhs.index) return 1;
			return 0;
		};
		const EmptyStat = function fnempty() {
			this.empty = 0;
			this.match = 0;
			this.nomatch = 0;
			this.total = 0;
		};
		const clear = () => {
			stats.length = 0;
			totals = new EmptyStat();
			stats[id.ALT] = new EmptyStat();
			stats[id.CAT] = new EmptyStat();
			stats[id.REP] = new EmptyStat();
			stats[id.RNM] = new EmptyStat();
			stats[id.TRG] = new EmptyStat();
			stats[id.TBS] = new EmptyStat();
			stats[id.TLS] = new EmptyStat();
			stats[id.UDT] = new EmptyStat();
			stats[id.AND] = new EmptyStat();
			stats[id.NOT] = new EmptyStat();
			ruleStats.length = 0;
			for (let i = 0; i < rules.length; i += 1) ruleStats.push({
				empty: 0,
				match: 0,
				nomatch: 0,
				total: 0,
				name: rules[i].name,
				lower: rules[i].lower,
				index: rules[i].index
			});
			if (udts.length > 0) {
				udtStats.length = 0;
				for (let i = 0; i < udts.length; i += 1) udtStats.push({
					empty: 0,
					match: 0,
					nomatch: 0,
					total: 0,
					name: udts[i].name,
					lower: udts[i].lower,
					index: udts[i].index
				});
			}
		};
		const incStat = (stat, state) => {
			stat.total += 1;
			switch (state) {
				case id.EMPTY:
					stat.empty += 1;
					break;
				case id.MATCH:
					stat.match += 1;
					break;
				case id.NOMATCH:
					stat.nomatch += 1;
					break;
				default: throw new Error(`${thisFileName}collect(): incStat(): unrecognized state: ${state}`);
			}
		};
	};
	var utilities = {
		stringToChars: (string) => [...string].map((cp) => cp.codePointAt(0)),
		charsToString: (chars, beg, len) => {
			let subChars = chars;
			while (1) {
				if (beg === void 0 || beg < 0) break;
				if (len === void 0) {
					subChars = chars.slice(beg);
					break;
				}
				if (len <= 0) return "";
				subChars = chars.slice(beg, beg + len);
				break;
			}
			return String.fromCodePoint(...subChars);
		}
	};
	var identifiers = {
		ALT: 1,
		CAT: 2,
		REP: 3,
		RNM: 4,
		TRG: 5,
		TBS: 6,
		TLS: 7,
		UDT: 11,
		AND: 12,
		NOT: 13,
		ACTIVE: 100,
		MATCH: 101,
		EMPTY: 102,
		NOMATCH: 103,
		SEM_PRE: 200,
		SEM_POST: 201,
		SEM_OK: 300,
		idName: (s) => {
			switch (s) {
				case identifiers.ALT: return "ALT";
				case identifiers.CAT: return "CAT";
				case identifiers.REP: return "REP";
				case identifiers.RNM: return "RNM";
				case identifiers.TRG: return "TRG";
				case identifiers.TBS: return "TBS";
				case identifiers.TLS: return "TLS";
				case identifiers.UDT: return "UDT";
				case identifiers.AND: return "AND";
				case identifiers.NOT: return "NOT";
				case identifiers.ACTIVE: return "ACTIVE";
				case identifiers.EMPTY: return "EMPTY";
				case identifiers.MATCH: return "MATCH";
				case identifiers.NOMATCH: return "NOMATCH";
				case identifiers.SEM_PRE: return "SEM_PRE";
				case identifiers.SEM_POST: return "SEM_POST";
				case identifiers.SEM_OK: return "SEM_OK";
				default: return "UNRECOGNIZED STATE";
			}
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/errors/JSONPointerError.mjs
	var JSONPointerError = class extends Error {
		constructor(message, options = void 0) {
			super(message, options);
			this.name = this.constructor.name;
			if (typeof message === "string") this.message = message;
			if (typeof Error.captureStackTrace === "function") Error.captureStackTrace(this, this.constructor);
			else this.stack = new Error(message).stack;
			/**
			* This needs to stay here until our minimum supported version of Node.js is >= 16.9.0.
			* Node.js is >= 16.9.0 supports error causes natively.
			*/
			if (options != null && typeof options === "object" && Object.prototype.hasOwnProperty.call(options, "cause") && !("cause" in this)) {
				const { cause } = options;
				this.cause = cause;
				if (cause instanceof Error && "stack" in cause) this.stack = `${this.stack}\nCAUSE: ${cause.stack}`;
			}
			/**
			* Allows to assign arbitrary properties to the error object.
			*/
			if (options != null && typeof options === "object") {
				const { cause, ...causelessOptions } = options;
				Object.assign(this, causelessOptions);
			}
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/errors/JSONPointerParseError.mjs
	var JSONPointerParseError = class extends JSONPointerError {};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/parse/callbacks/cst.mjs
	var cst = (ruleName) => {
		return (state, chars, phraseIndex, phraseLength, data) => {
			if (!(typeof data === "object" && data !== null && !Array.isArray(data))) throw new JSONPointerParseError("parser's user data must be an object");
			if (state === identifiers.SEM_PRE) {
				const node = {
					type: ruleName,
					text: utilities.charsToString(chars, phraseIndex, phraseLength),
					start: phraseIndex,
					length: phraseLength,
					children: []
				};
				if (data.stack.length > 0) data.stack[data.stack.length - 1].children.push(node);
				else data.root = node;
				data.stack.push(node);
			}
			if (state === identifiers.SEM_POST) data.stack.pop();
		};
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/parse/translators/CSTTranslator.mjs
	var CSTTranslator = class extends Ast {
		constructor() {
			super();
			this.callbacks["json-pointer"] = cst("json-pointer");
			this.callbacks["reference-token"] = cst("reference-token");
			this.callbacks["slash"] = cst("text");
		}
		getTree() {
			const data = {
				stack: [],
				root: null
			};
			this.translate(data);
			delete data.stack;
			return data;
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/unescape.mjs
	var unescape = (referenceToken) => {
		if (typeof referenceToken !== "string") throw new TypeError("Reference token must be a string");
		return referenceToken.replace(/~1/g, "/").replace(/~0/g, "~");
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/parse/translators/ASTTranslator.mjs
	var ASTTranslator = class extends CSTTranslator {
		getTree() {
			const { root } = super.getTree();
			return root.children.filter(({ type }) => type === "reference-token").map(({ text }) => unescape(text));
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/parse/trace/Expectations.mjs
	var Expectations = class extends Array {
		toString() {
			return this.map((c) => `"${String(c)}"`).join(", ");
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/parse/trace/Trace.mjs
	var Trace = class extends Trace$1 {
		inferExpectations() {
			const lines = this.displayTrace().split("\n");
			const expectations = /* @__PURE__ */ new Set();
			let lastMatchedIndex = -1;
			for (let i = 0; i < lines.length; i++) {
				const line = lines[i];
				if (line.includes("M|")) {
					const textMatch = line.match(/]'(.*)'$/);
					if (textMatch && textMatch[1]) lastMatchedIndex = i;
				}
				if (i > lastMatchedIndex) {
					const terminalFailMatch = line.match(/N\|\[TLS\(([^)]+)\)]/);
					if (terminalFailMatch) expectations.add(terminalFailMatch[1]);
				}
			}
			return new Expectations(...expectations);
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/parse/index.mjs
	var grammar$5 = new grammar();
	var parse = (jsonPointer, { translator = new ASTTranslator(), stats = false, trace = false } = {}) => {
		if (typeof jsonPointer !== "string") throw new TypeError("JSON Pointer must be a string");
		try {
			const parser = new Parser();
			if (translator) parser.ast = translator;
			if (stats) parser.stats = new Stats();
			if (trace) parser.trace = new Trace();
			const result = parser.parse(grammar$5, "json-pointer", jsonPointer);
			return {
				result,
				tree: result.success && translator ? parser.ast.getTree() : void 0,
				stats: parser.stats,
				trace: parser.trace
			};
		} catch (error) {
			throw new JSONPointerParseError("Unexpected error during JSON Pointer parsing", {
				cause: error,
				jsonPointer
			});
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/parse/translators/XMLTranslator.mjs
	var XMLTranslator = class extends CSTTranslator {
		getTree() {
			return this.toXml();
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/test/json-pointer.mjs
	var testJSONPointer = (jsonPointer) => {
		try {
			return parse(jsonPointer, { translator: null }).result.success;
		} catch {
			return false;
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/test/reference-token.mjs
	var grammar$4 = new grammar();
	var parser$3 = new Parser();
	var testReferenceToken = (referenceToken) => {
		if (typeof referenceToken !== "string") return false;
		try {
			return parser$3.parse(grammar$4, "reference-token", referenceToken).success;
		} catch {
			return false;
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/test/array-location.mjs
	var grammar$3 = new grammar();
	var parser$2 = new Parser();
	var testArrayLocation = (referenceToken) => {
		if (typeof referenceToken !== "string") return false;
		try {
			return parser$2.parse(grammar$3, "array-location", referenceToken).success;
		} catch {
			return false;
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/test/array-index.mjs
	var grammar$2 = new grammar();
	var parser$1 = new Parser();
	var testArrayIndex = (referenceToken) => {
		if (typeof referenceToken !== "string") return false;
		try {
			return parser$1.parse(grammar$2, "array-index", referenceToken).success;
		} catch {
			return false;
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/test/array-dash.mjs
	var grammar$1 = new grammar();
	var parser = new Parser();
	var testArrayDash = (referenceToken) => {
		if (typeof referenceToken !== "string") return false;
		try {
			return parser.parse(grammar$1, "array-dash", referenceToken).success;
		} catch {
			return false;
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/escape.mjs
	var escape = (referenceToken) => {
		if (typeof referenceToken !== "string" && typeof referenceToken !== "number") throw new TypeError("Reference token must be a string or number");
		return String(referenceToken).replace(/~/g, "~0").replace(/\//g, "~1");
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/errors/JSONPointerCompileError.mjs
	var JSONPointerCompileError = class extends JSONPointerError {};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/compile.mjs
	var compile = (referenceTokens) => {
		if (!Array.isArray(referenceTokens)) throw new TypeError("Reference tokens must be a list of strings or numbers");
		try {
			if (referenceTokens.length === 0) return "";
			return `/${referenceTokens.map((referenceToken) => {
				if (typeof referenceToken !== "string" && typeof referenceToken !== "number") throw new TypeError("Reference token must be a string or number");
				return escape(String(referenceToken));
			}).join("/")}`;
		} catch (error) {
			throw new JSONPointerCompileError("Unexpected error during JSON Pointer compilation", {
				cause: error,
				referenceTokens
			});
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/evaluate/trace/TraceBuilder.mjs
	var TraceBuilder = class {
		#trace;
		#path;
		#realm;
		constructor(trace, context = {}) {
			this.#trace = trace;
			this.#trace.steps = [];
			this.#trace.failed = false;
			this.#trace.failedAt = -1;
			this.#trace.message = `JSON Pointer "${context.jsonPointer}" was successfully evaluated against the provided value`;
			this.#trace.context = {
				...context,
				realm: context.realm.name
			};
			this.#path = [];
			this.#realm = context.realm;
		}
		step({ referenceToken, input, output, success = true, reason }) {
			const position = this.#path.length;
			this.#path.push(referenceToken);
			const step = {
				referenceToken,
				referenceTokenPosition: position,
				input,
				inputType: this.#realm.isObject(input) ? "object" : this.#realm.isArray(input) ? "array" : "unrecognized",
				output,
				success
			};
			if (reason) step.reason = reason;
			this.#trace.steps.push(step);
			if (!success) {
				this.#trace.failed = true;
				this.#trace.failedAt = position;
				this.#trace.message = reason;
			}
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/evaluate/realms/EvaluationRealm.mjs
	var EvaluationRealm = class {
		name = "";
		isArray(node) {
			throw new JSONPointerError("Realm.isArray(node) must be implemented in a subclass");
		}
		isObject(node) {
			throw new JSONPointerError("Realm.isObject(node) must be implemented in a subclass");
		}
		sizeOf(node) {
			throw new JSONPointerError("Realm.sizeOf(node) must be implemented in a subclass");
		}
		has(node, referenceToken) {
			throw new JSONPointerError("Realm.has(node) must be implemented in a subclass");
		}
		evaluate(node, referenceToken) {
			throw new JSONPointerError("Realm.evaluate(node) must be implemented in a subclass");
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/errors/JSONPointerEvaluateError.mjs
	var JSONPointerEvaluateError = class extends JSONPointerError {};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/errors/JSONPointerIndexError.mjs
	var JSONPointerIndexError = class extends JSONPointerEvaluateError {};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/evaluate/realms/json/index.mjs
	var JSONEvaluationRealm = class extends EvaluationRealm {
		name = "json";
		isArray(node) {
			return Array.isArray(node);
		}
		isObject(node) {
			return typeof node === "object" && node !== null && !this.isArray(node);
		}
		sizeOf(node) {
			if (this.isArray(node)) return node.length;
			if (this.isObject(node)) return Object.keys(node).length;
			return 0;
		}
		has(node, referenceToken) {
			if (this.isArray(node)) {
				const index = Number(referenceToken);
				const indexUint32 = index >>> 0;
				if (index !== indexUint32) throw new JSONPointerIndexError(`Invalid array index "${referenceToken}": index must be an unsinged 32-bit integer`, {
					referenceToken,
					currentValue: node,
					realm: this.name
				});
				return indexUint32 < this.sizeOf(node) && Object.prototype.hasOwnProperty.call(node, index);
			}
			if (this.isObject(node)) return Object.prototype.hasOwnProperty.call(node, referenceToken);
			return false;
		}
		evaluate(node, referenceToken) {
			if (this.isArray(node)) return node[Number(referenceToken)];
			return node[referenceToken];
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/errors/JSONPointerTypeError.mjs
	var JSONPointerTypeError = class extends JSONPointerEvaluateError {};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/errors/JSONPointerKeyError.mjs
	var JSONPointerKeyError = class extends JSONPointerEvaluateError {};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/evaluate/index.mjs
	var evaluate$1 = (value, jsonPointer, { strictArrays = true, strictObjects = true, realm = new JSONEvaluationRealm(), trace = true } = {}) => {
		const { result: parseResult, tree: referenceTokens, trace: parseTrace } = parse(jsonPointer, { trace: !!trace });
		const tracer = typeof trace === "object" && trace !== null ? new TraceBuilder(trace, {
			jsonPointer,
			referenceTokens,
			strictArrays,
			strictObjects,
			realm,
			value
		}) : null;
		try {
			let output;
			if (!parseResult.success) {
				let message = `Invalid JSON Pointer: "${jsonPointer}". Syntax error at position ${parseResult.maxMatched}`;
				message += parseTrace ? `, expected ${parseTrace.inferExpectations()}` : "";
				throw new JSONPointerEvaluateError(message, {
					jsonPointer,
					currentValue: value,
					realm: realm.name
				});
			}
			return referenceTokens.reduce((current, referenceToken, referenceTokenPosition) => {
				if (realm.isArray(current)) {
					if (testArrayDash(referenceToken)) if (strictArrays) throw new JSONPointerIndexError(`Invalid array index "-" at position ${referenceTokenPosition} in "${jsonPointer}". The "-" token always refers to a nonexistent element during evaluation`, {
						jsonPointer,
						referenceTokens,
						referenceToken,
						referenceTokenPosition,
						currentValue: current,
						realm: realm.name
					});
					else {
						output = realm.evaluate(current, String(realm.sizeOf(current)));
						tracer === null || tracer === void 0 || tracer.step({
							referenceToken,
							input: current,
							output
						});
						return output;
					}
					if (!testArrayIndex(referenceToken)) throw new JSONPointerIndexError(`Invalid array index "${referenceToken}" at position ${referenceTokenPosition} in "${jsonPointer}": index MUST be "0", or digits without a leading "0"`, {
						jsonPointer,
						referenceTokens,
						referenceToken,
						referenceTokenPosition,
						currentValue: current,
						realm: realm.name
					});
					const index = Number(referenceToken);
					if (!Number.isSafeInteger(index)) throw new JSONPointerIndexError(`Invalid array index "${referenceToken}" at position ${referenceTokenPosition} in "${jsonPointer}": index must be a safe integer`, {
						jsonPointer,
						referenceTokens,
						referenceToken,
						referenceTokenPosition,
						currentValue: current,
						realm: realm.name
					});
					if (!realm.has(current, referenceToken) && strictArrays) throw new JSONPointerIndexError(`Invalid array index "${referenceToken}" at position ${referenceTokenPosition} in "${jsonPointer}": index not found in array`, {
						jsonPointer,
						referenceTokens,
						referenceToken,
						referenceTokenPosition,
						currentValue: current,
						realm: realm.name
					});
					output = realm.evaluate(current, referenceToken);
					tracer === null || tracer === void 0 || tracer.step({
						referenceToken,
						input: current,
						output
					});
					return output;
				}
				if (realm.isObject(current)) {
					if (!realm.has(current, referenceToken) && strictObjects) throw new JSONPointerKeyError(`Invalid object key "${referenceToken}" at position ${referenceTokenPosition} in "${jsonPointer}": key not found in object`, {
						jsonPointer,
						referenceTokens,
						referenceToken,
						referenceTokenPosition,
						currentValue: current,
						realm: realm.name
					});
					output = realm.evaluate(current, referenceToken);
					tracer === null || tracer === void 0 || tracer.step({
						referenceToken,
						input: current,
						output
					});
					return output;
				}
				throw new JSONPointerTypeError(`Invalid reference token "${referenceToken}" at position ${referenceTokenPosition} in "${jsonPointer}": cannot be applied to a non-object/non-array value`, {
					jsonPointer,
					referenceTokens,
					referenceToken,
					referenceTokenPosition,
					currentValue: current,
					realm: realm.name
				});
			}, value);
		} catch (error) {
			tracer === null || tracer === void 0 || tracer.step({
				referenceToken: error.referenceToken,
				input: error.currentValue,
				success: false,
				reason: error.message
			});
			if (error instanceof JSONPointerEvaluateError) throw error;
			throw new JSONPointerEvaluateError("Unexpected error during JSON Pointer evaluation", {
				cause: error,
				jsonPointer,
				referenceTokens
			});
		}
	};
	//#endregion
	//#region ../../node_modules/ramda/es/internal/_isPlaceholder.js
	function _isPlaceholder(a) {
		return a != null && typeof a === "object" && a["@@functional/placeholder"] === true;
	}
	//#endregion
	//#region ../../node_modules/ramda/es/internal/_curry1.js
	/**
	* Optimized internal one-arity curry function.
	*
	* @private
	* @category Function
	* @param {Function} fn The function to curry.
	* @return {Function} The curried function.
	*/
	function _curry1(fn) {
		return function f1(a) {
			if (arguments.length === 0 || _isPlaceholder(a)) return f1;
			else return fn.apply(this, arguments);
		};
	}
	//#endregion
	//#region ../../node_modules/ramda/es/internal/_curry2.js
	/**
	* Optimized internal two-arity curry function.
	*
	* @private
	* @category Function
	* @param {Function} fn The function to curry.
	* @return {Function} The curried function.
	*/
	function _curry2(fn) {
		return function f2(a, b) {
			switch (arguments.length) {
				case 0: return f2;
				case 1: return _isPlaceholder(a) ? f2 : _curry1(function(_b) {
					return fn(a, _b);
				});
				default: return _isPlaceholder(a) && _isPlaceholder(b) ? f2 : _isPlaceholder(a) ? _curry1(function(_a) {
					return fn(_a, b);
				}) : _isPlaceholder(b) ? _curry1(function(_b) {
					return fn(a, _b);
				}) : fn(a, b);
			}
		};
	}
	//#endregion
	//#region ../../node_modules/ramda/es/internal/_arrayFromIterator.js
	function _arrayFromIterator(iter) {
		var list = [];
		var next;
		while (!(next = iter.next()).done) list.push(next.value);
		return list;
	}
	//#endregion
	//#region ../../node_modules/ramda/es/internal/_includesWith.js
	function _includesWith(pred, x, list) {
		var idx = 0;
		var len = list.length;
		while (idx < len) {
			if (pred(x, list[idx])) return true;
			idx += 1;
		}
		return false;
	}
	//#endregion
	//#region ../../node_modules/ramda/es/internal/_functionName.js
	function _functionName(f) {
		var match = String(f).match(/^function (\w*)/);
		return match == null ? "" : match[1];
	}
	//#endregion
	//#region ../../node_modules/ramda/es/internal/_has.js
	function _has(prop, obj) {
		return Object.prototype.hasOwnProperty.call(obj, prop);
	}
	//#endregion
	//#region ../../node_modules/ramda/es/internal/_objectIs.js
	function _objectIs(a, b) {
		if (a === b) return a !== 0 || 1 / a === 1 / b;
		else return a !== a && b !== b;
	}
	var _objectIs_default = typeof Object.is === "function" ? Object.is : _objectIs;
	//#endregion
	//#region ../../node_modules/ramda/es/internal/_isArguments.js
	var toString = Object.prototype.toString;
	var _isArguments = /*#__PURE__*/ function() {
		return toString.call(arguments) === "[object Arguments]" ? function _isArguments(x) {
			return toString.call(x) === "[object Arguments]";
		} : function _isArguments(x) {
			return _has("callee", x);
		};
	}();
	//#endregion
	//#region ../../node_modules/ramda/es/keys.js
	var hasEnumBug = !/*#__PURE__*/ { toString: null }.propertyIsEnumerable("toString");
	var nonEnumerableProps = [
		"constructor",
		"valueOf",
		"isPrototypeOf",
		"toString",
		"propertyIsEnumerable",
		"hasOwnProperty",
		"toLocaleString"
	];
	var hasArgsEnumBug = /*#__PURE__*/ function() {
		"use strict";
		return arguments.propertyIsEnumerable("length");
	}();
	var contains = function contains(list, item) {
		var idx = 0;
		while (idx < list.length) {
			if (list[idx] === item) return true;
			idx += 1;
		}
		return false;
	};
	/**
	* Returns a list containing the names of all the enumerable own properties of
	* the supplied object.
	* Note that the order of the output array is not guaranteed to be consistent
	* across different JS platforms.
	*
	* @func
	* @memberOf R
	* @since v0.1.0
	* @category Object
	* @sig {k: v} -> [k]
	* @param {Object} obj The object to extract properties from
	* @return {Array} An array of the object's own properties.
	* @see R.keysIn, R.values, R.toPairs
	* @example
	*
	*      R.keys({a: 1, b: 2, c: 3}); //=> ['a', 'b', 'c']
	*/
	var keys = typeof Object.keys === "function" && !hasArgsEnumBug ? /*#__PURE__*/ _curry1(function keys(obj) {
		return Object(obj) !== obj ? [] : Object.keys(obj);
	}) : /*#__PURE__*/ _curry1(function keys(obj) {
		if (Object(obj) !== obj) return [];
		var prop, nIdx;
		var ks = [];
		var checkArgsLength = hasArgsEnumBug && _isArguments(obj);
		for (prop in obj) if (_has(prop, obj) && (!checkArgsLength || prop !== "length")) ks[ks.length] = prop;
		if (hasEnumBug) {
			nIdx = nonEnumerableProps.length - 1;
			while (nIdx >= 0) {
				prop = nonEnumerableProps[nIdx];
				if (_has(prop, obj) && !contains(ks, prop)) ks[ks.length] = prop;
				nIdx -= 1;
			}
		}
		return ks;
	});
	//#endregion
	//#region ../../node_modules/ramda/es/type.js
	/**
	* Gives a single-word string description of the (native) type of a value,
	* returning such answers as 'Object', 'Number', 'Array', or 'Null'. Does not
	* attempt to distinguish user Object types any further, reporting them all as
	* 'Object'.
	*
	* @func
	* @memberOf R
	* @since v0.8.0
	* @category Type
	* @sig * -> String
	* @param {*} val The value to test
	* @return {String}
	* @example
	*
	*      R.type({}); //=> "Object"
	*      R.type(1); //=> "Number"
	*      R.type(false); //=> "Boolean"
	*      R.type('s'); //=> "String"
	*      R.type(null); //=> "Null"
	*      R.type([]); //=> "Array"
	*      R.type(/[A-z]/); //=> "RegExp"
	*      R.type(() => {}); //=> "Function"
	*      R.type(async () => {}); //=> "AsyncFunction"
	*      R.type(undefined); //=> "Undefined"
	*      R.type(BigInt(123)); //=> "BigInt"
	*/
	var type = /*#__PURE__*/ _curry1(function type(val) {
		return val === null ? "Null" : val === void 0 ? "Undefined" : Object.prototype.toString.call(val).slice(8, -1);
	});
	//#endregion
	//#region ../../node_modules/ramda/es/internal/_equals.js
	/**
	* private _uniqContentEquals function.
	* That function is checking equality of 2 iterator contents with 2 assumptions
	* - iterators lengths are the same
	* - iterators values are unique
	*
	* false-positive result will be returned for comparison of, e.g.
	* - [1,2,3] and [1,2,3,4]
	* - [1,1,1] and [1,2,3]
	* */
	function _uniqContentEquals(aIterator, bIterator, stackA, stackB) {
		var a = _arrayFromIterator(aIterator);
		var b = _arrayFromIterator(bIterator);
		function eq(_a, _b) {
			return _equals(_a, _b, stackA.slice(), stackB.slice());
		}
		return !_includesWith(function(b, aItem) {
			return !_includesWith(eq, aItem, b);
		}, b, a);
	}
	function _equals(a, b, stackA, stackB) {
		if (_objectIs_default(a, b)) return true;
		var typeA = type(a);
		if (typeA !== type(b)) return false;
		if (typeof a["fantasy-land/equals"] === "function" || typeof b["fantasy-land/equals"] === "function") return typeof a["fantasy-land/equals"] === "function" && a["fantasy-land/equals"](b) && typeof b["fantasy-land/equals"] === "function" && b["fantasy-land/equals"](a);
		if (typeof a.equals === "function" || typeof b.equals === "function") return typeof a.equals === "function" && a.equals(b) && typeof b.equals === "function" && b.equals(a);
		switch (typeA) {
			case "Arguments":
			case "Array":
			case "Object":
				if (typeof a.constructor === "function" && _functionName(a.constructor) === "Promise") return a === b;
				break;
			case "Boolean":
			case "Number":
			case "String":
				if (!(typeof a === typeof b && _objectIs_default(a.valueOf(), b.valueOf()))) return false;
				break;
			case "Date":
				if (!_objectIs_default(a.valueOf(), b.valueOf())) return false;
				break;
			case "Error": return a.name === b.name && a.message === b.message;
			case "RegExp":
				if (!(a.source === b.source && a.global === b.global && a.ignoreCase === b.ignoreCase && a.multiline === b.multiline && a.sticky === b.sticky && a.unicode === b.unicode)) return false;
				break;
		}
		var idx = stackA.length - 1;
		while (idx >= 0) {
			if (stackA[idx] === a) return stackB[idx] === b;
			idx -= 1;
		}
		switch (typeA) {
			case "Map":
				if (a.size !== b.size) return false;
				return _uniqContentEquals(a.entries(), b.entries(), stackA.concat([a]), stackB.concat([b]));
			case "Set":
				if (a.size !== b.size) return false;
				return _uniqContentEquals(a.values(), b.values(), stackA.concat([a]), stackB.concat([b]));
			case "Arguments":
			case "Array":
			case "Object":
			case "Boolean":
			case "Number":
			case "String":
			case "Date":
			case "Error":
			case "RegExp":
			case "Int8Array":
			case "Uint8Array":
			case "Uint8ClampedArray":
			case "Int16Array":
			case "Uint16Array":
			case "Int32Array":
			case "Uint32Array":
			case "Float32Array":
			case "Float64Array":
			case "ArrayBuffer": break;
			default: return false;
		}
		var keysA = keys(a);
		if (keysA.length !== keys(b).length) return false;
		var extendedStackA = stackA.concat([a]);
		var extendedStackB = stackB.concat([b]);
		idx = keysA.length - 1;
		while (idx >= 0) {
			var key = keysA[idx];
			if (!(_has(key, b) && _equals(b[key], a[key], extendedStackA, extendedStackB))) return false;
			idx -= 1;
		}
		return true;
	}
	//#endregion
	//#region ../../node_modules/ramda-adjunct/es/isUndefined.js
	/**
	* Checks if input value is `undefined`.
	*
	* @func isUndefined
	* @memberOf RA
	* @since {@link https://char0n.github.io/ramda-adjunct/0.0.1|v0.0.1}
	* @category Type
	* @sig * -> Boolean
	* @param {*} val The value to test
	* @return {boolean}
	* @see {@link RA.isNotUndefined|isNotUndefined}
	* @example
	*
	* RA.isUndefined(1); //=> false
	* RA.isUndefined(undefined); //=> true
	* RA.isUndefined(null); //=> false
	*/
	var isUndefined = (/* @__PURE__ */ _curry2(function equals(a, b) {
		return _equals(a, b, [], []);
	}))((/* @__PURE__ */ _curry1(function always(val) {
		return function() {
			return val;
		};
	}))(void 0)());
	//#endregion
	//#region ../../node_modules/lodash/isNull.js
	var require_isNull = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Checks if `value` is `null`.
		*
		* @static
		* @memberOf _
		* @since 0.1.0
		* @category Lang
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is `null`, else `false`.
		* @example
		*
		* _.isNull(null);
		* // => true
		*
		* _.isNull(void 0);
		* // => false
		*/
		function isNull(value) {
			return value === null;
		}
		module.exports = isNull;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_freeGlobal.js
	var require__freeGlobal = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = typeof global == "object" && global && global.Object === Object && global;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_root.js
	var require__root = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var freeGlobal = require__freeGlobal();
		/** Detect free variable `self`. */
		var freeSelf = typeof self == "object" && self && self.Object === Object && self;
		module.exports = freeGlobal || freeSelf || Function("return this")();
	}));
	//#endregion
	//#region ../../node_modules/lodash/_Symbol.js
	var require__Symbol = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require__root().Symbol;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_getRawTag.js
	var require__getRawTag = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Symbol = require__Symbol();
		/** Used for built-in method references. */
		var objectProto = Object.prototype;
		/** Used to check objects for own properties. */
		var hasOwnProperty = objectProto.hasOwnProperty;
		/**
		* Used to resolve the
		* [`toStringTag`](http://ecma-international.org/ecma-262/7.0/#sec-object.prototype.tostring)
		* of values.
		*/
		var nativeObjectToString = objectProto.toString;
		/** Built-in value references. */
		var symToStringTag = Symbol ? Symbol.toStringTag : void 0;
		/**
		* A specialized version of `baseGetTag` which ignores `Symbol.toStringTag` values.
		*
		* @private
		* @param {*} value The value to query.
		* @returns {string} Returns the raw `toStringTag`.
		*/
		function getRawTag(value) {
			var isOwn = hasOwnProperty.call(value, symToStringTag), tag = value[symToStringTag];
			try {
				value[symToStringTag] = void 0;
				var unmasked = true;
			} catch (e) {}
			var result = nativeObjectToString.call(value);
			if (unmasked) if (isOwn) value[symToStringTag] = tag;
			else delete value[symToStringTag];
			return result;
		}
		module.exports = getRawTag;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_objectToString.js
	var require__objectToString = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Used to resolve the
		* [`toStringTag`](http://ecma-international.org/ecma-262/7.0/#sec-object.prototype.tostring)
		* of values.
		*/
		var nativeObjectToString = Object.prototype.toString;
		/**
		* Converts `value` to a string using `Object.prototype.toString`.
		*
		* @private
		* @param {*} value The value to convert.
		* @returns {string} Returns the converted string.
		*/
		function objectToString(value) {
			return nativeObjectToString.call(value);
		}
		module.exports = objectToString;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_baseGetTag.js
	var require__baseGetTag = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Symbol = require__Symbol(), getRawTag = require__getRawTag(), objectToString = require__objectToString();
		/** `Object#toString` result references. */
		var nullTag = "[object Null]", undefinedTag = "[object Undefined]";
		/** Built-in value references. */
		var symToStringTag = Symbol ? Symbol.toStringTag : void 0;
		/**
		* The base implementation of `getTag` without fallbacks for buggy environments.
		*
		* @private
		* @param {*} value The value to query.
		* @returns {string} Returns the `toStringTag`.
		*/
		function baseGetTag(value) {
			if (value == null) return value === void 0 ? undefinedTag : nullTag;
			return symToStringTag && symToStringTag in Object(value) ? getRawTag(value) : objectToString(value);
		}
		module.exports = baseGetTag;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isArray.js
	var require_isArray = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = Array.isArray;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isObjectLike.js
	var require_isObjectLike = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Checks if `value` is object-like. A value is object-like if it's not `null`
		* and has a `typeof` result of "object".
		*
		* @static
		* @memberOf _
		* @since 4.0.0
		* @category Lang
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is object-like, else `false`.
		* @example
		*
		* _.isObjectLike({});
		* // => true
		*
		* _.isObjectLike([1, 2, 3]);
		* // => true
		*
		* _.isObjectLike(_.noop);
		* // => false
		*
		* _.isObjectLike(null);
		* // => false
		*/
		function isObjectLike(value) {
			return value != null && typeof value == "object";
		}
		module.exports = isObjectLike;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isString.js
	var require_isString = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseGetTag = require__baseGetTag(), isArray = require_isArray(), isObjectLike = require_isObjectLike();
		/** `Object#toString` result references. */
		var stringTag = "[object String]";
		/**
		* Checks if `value` is classified as a `String` primitive or object.
		*
		* @static
		* @since 0.1.0
		* @memberOf _
		* @category Lang
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is a string, else `false`.
		* @example
		*
		* _.isString('abc');
		* // => true
		*
		* _.isString(1);
		* // => false
		*/
		function isString(value) {
			return typeof value == "string" || !isArray(value) && isObjectLike(value) && baseGetTag(value) == stringTag;
		}
		module.exports = isString;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isNumber.js
	var require_isNumber = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseGetTag = require__baseGetTag(), isObjectLike = require_isObjectLike();
		/** `Object#toString` result references. */
		var numberTag = "[object Number]";
		/**
		* Checks if `value` is classified as a `Number` primitive or object.
		*
		* **Note:** To exclude `Infinity`, `-Infinity`, and `NaN`, which are
		* classified as numbers, use the `_.isFinite` method.
		*
		* @static
		* @memberOf _
		* @since 0.1.0
		* @category Lang
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is a number, else `false`.
		* @example
		*
		* _.isNumber(3);
		* // => true
		*
		* _.isNumber(Number.MIN_VALUE);
		* // => true
		*
		* _.isNumber(Infinity);
		* // => true
		*
		* _.isNumber('3');
		* // => false
		*/
		function isNumber(value) {
			return typeof value == "number" || isObjectLike(value) && baseGetTag(value) == numberTag;
		}
		module.exports = isNumber;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isBoolean.js
	var require_isBoolean = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseGetTag = require__baseGetTag(), isObjectLike = require_isObjectLike();
		/** `Object#toString` result references. */
		var boolTag = "[object Boolean]";
		/**
		* Checks if `value` is classified as a boolean primitive or object.
		*
		* @static
		* @memberOf _
		* @since 0.1.0
		* @category Lang
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is a boolean, else `false`.
		* @example
		*
		* _.isBoolean(false);
		* // => true
		*
		* _.isBoolean(null);
		* // => false
		*/
		function isBoolean(value) {
			return value === true || value === false || isObjectLike(value) && baseGetTag(value) == boolTag;
		}
		module.exports = isBoolean;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isObject.js
	var require_isObject = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Checks if `value` is the
		* [language type](http://www.ecma-international.org/ecma-262/7.0/#sec-ecmascript-language-types)
		* of `Object`. (e.g. arrays, functions, objects, regexes, `new Number(0)`, and `new String('')`)
		*
		* @static
		* @memberOf _
		* @since 0.1.0
		* @category Lang
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is an object, else `false`.
		* @example
		*
		* _.isObject({});
		* // => true
		*
		* _.isObject([1, 2, 3]);
		* // => true
		*
		* _.isObject(_.noop);
		* // => true
		*
		* _.isObject(null);
		* // => false
		*/
		function isObject(value) {
			var type = typeof value;
			return value != null && (type == "object" || type == "function");
		}
		module.exports = isObject;
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/serialisers/JSONSerialiser.js
	var require_JSONSerialiser = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* @class JSONSerialiser
		*
		* @param {Namespace} namespace
		*
		* @property {Namespace} namespace
		*/
		var JSONSerialiser = class {
			constructor(namespace) {
				this.namespace = namespace || new this.Namespace();
			}
			/**
			* @param {Element} element
			* @returns {object}
			*/
			serialise(element) {
				if (!(element instanceof this.namespace.elements.Element)) throw new TypeError(`Given element \`${element}\` is not an Element instance`);
				const payload = { element: element.element };
				if (element._meta && element._meta.length > 0) payload.meta = this.serialiseObject(element.meta);
				if (element._attributes && element._attributes.length > 0) payload.attributes = this.serialiseObject(element.attributes);
				const content = this.serialiseContent(element.content);
				if (content !== void 0) payload.content = content;
				return payload;
			}
			/**
			* @param {object} value
			* @returns {Element}
			*/
			deserialise(value) {
				if (!value.element) throw new Error("Given value is not an object containing an element name");
				const element = new (this.namespace.getElementClass(value.element))();
				if (element.element !== value.element) element.element = value.element;
				if (value.meta) this.deserialiseObject(value.meta, element.meta);
				if (value.attributes) this.deserialiseObject(value.attributes, element.attributes);
				const content = this.deserialiseContent(value.content);
				if (content !== void 0 || element.content === null) element.content = content;
				return element;
			}
			serialiseContent(content) {
				if (content instanceof this.namespace.elements.Element) return this.serialise(content);
				if (content instanceof this.namespace.KeyValuePair) {
					const pair = { key: this.serialise(content.key) };
					if (content.value) pair.value = this.serialise(content.value);
					return pair;
				}
				if (content && content.map) {
					if (content.length === 0) return;
					return content.map(this.serialise, this);
				}
				return content;
			}
			deserialiseContent(content) {
				if (content) {
					if (content.element) return this.deserialise(content);
					if (content.key) {
						const pair = new this.namespace.KeyValuePair(this.deserialise(content.key));
						if (content.value) pair.value = this.deserialise(content.value);
						return pair;
					}
					if (content.map) return content.map(this.deserialise, this);
				}
				return content;
			}
			serialiseObject(obj) {
				const result = {};
				obj.forEach((value, key) => {
					if (value) result[key.toValue()] = this.serialise(value);
				});
				if (Object.keys(result).length === 0) return;
				return result;
			}
			deserialiseObject(from, to) {
				Object.keys(from).forEach((key) => {
					to.set(key, this.deserialise(from[key]));
				});
			}
		};
		module.exports = JSONSerialiser;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_listCacheClear.js
	var require__listCacheClear = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Removes all key-value entries from the list cache.
		*
		* @private
		* @name clear
		* @memberOf ListCache
		*/
		function listCacheClear() {
			this.__data__ = [];
			this.size = 0;
		}
		module.exports = listCacheClear;
	}));
	//#endregion
	//#region ../../node_modules/lodash/eq.js
	var require_eq = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Performs a
		* [`SameValueZero`](http://ecma-international.org/ecma-262/7.0/#sec-samevaluezero)
		* comparison between two values to determine if they are equivalent.
		*
		* @static
		* @memberOf _
		* @since 4.0.0
		* @category Lang
		* @param {*} value The value to compare.
		* @param {*} other The other value to compare.
		* @returns {boolean} Returns `true` if the values are equivalent, else `false`.
		* @example
		*
		* var object = { 'a': 1 };
		* var other = { 'a': 1 };
		*
		* _.eq(object, object);
		* // => true
		*
		* _.eq(object, other);
		* // => false
		*
		* _.eq('a', 'a');
		* // => true
		*
		* _.eq('a', Object('a'));
		* // => false
		*
		* _.eq(NaN, NaN);
		* // => true
		*/
		function eq(value, other) {
			return value === other || value !== value && other !== other;
		}
		module.exports = eq;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_assocIndexOf.js
	var require__assocIndexOf = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var eq = require_eq();
		/**
		* Gets the index at which the `key` is found in `array` of key-value pairs.
		*
		* @private
		* @param {Array} array The array to inspect.
		* @param {*} key The key to search for.
		* @returns {number} Returns the index of the matched value, else `-1`.
		*/
		function assocIndexOf(array, key) {
			var length = array.length;
			while (length--) if (eq(array[length][0], key)) return length;
			return -1;
		}
		module.exports = assocIndexOf;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_listCacheDelete.js
	var require__listCacheDelete = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var assocIndexOf = require__assocIndexOf();
		/** Built-in value references. */
		var splice = Array.prototype.splice;
		/**
		* Removes `key` and its value from the list cache.
		*
		* @private
		* @name delete
		* @memberOf ListCache
		* @param {string} key The key of the value to remove.
		* @returns {boolean} Returns `true` if the entry was removed, else `false`.
		*/
		function listCacheDelete(key) {
			var data = this.__data__, index = assocIndexOf(data, key);
			if (index < 0) return false;
			if (index == data.length - 1) data.pop();
			else splice.call(data, index, 1);
			--this.size;
			return true;
		}
		module.exports = listCacheDelete;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_listCacheGet.js
	var require__listCacheGet = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var assocIndexOf = require__assocIndexOf();
		/**
		* Gets the list cache value for `key`.
		*
		* @private
		* @name get
		* @memberOf ListCache
		* @param {string} key The key of the value to get.
		* @returns {*} Returns the entry value.
		*/
		function listCacheGet(key) {
			var data = this.__data__, index = assocIndexOf(data, key);
			return index < 0 ? void 0 : data[index][1];
		}
		module.exports = listCacheGet;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_listCacheHas.js
	var require__listCacheHas = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var assocIndexOf = require__assocIndexOf();
		/**
		* Checks if a list cache value for `key` exists.
		*
		* @private
		* @name has
		* @memberOf ListCache
		* @param {string} key The key of the entry to check.
		* @returns {boolean} Returns `true` if an entry for `key` exists, else `false`.
		*/
		function listCacheHas(key) {
			return assocIndexOf(this.__data__, key) > -1;
		}
		module.exports = listCacheHas;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_listCacheSet.js
	var require__listCacheSet = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var assocIndexOf = require__assocIndexOf();
		/**
		* Sets the list cache `key` to `value`.
		*
		* @private
		* @name set
		* @memberOf ListCache
		* @param {string} key The key of the value to set.
		* @param {*} value The value to set.
		* @returns {Object} Returns the list cache instance.
		*/
		function listCacheSet(key, value) {
			var data = this.__data__, index = assocIndexOf(data, key);
			if (index < 0) {
				++this.size;
				data.push([key, value]);
			} else data[index][1] = value;
			return this;
		}
		module.exports = listCacheSet;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_ListCache.js
	var require__ListCache = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var listCacheClear = require__listCacheClear(), listCacheDelete = require__listCacheDelete(), listCacheGet = require__listCacheGet(), listCacheHas = require__listCacheHas(), listCacheSet = require__listCacheSet();
		/**
		* Creates an list cache object.
		*
		* @private
		* @constructor
		* @param {Array} [entries] The key-value pairs to cache.
		*/
		function ListCache(entries) {
			var index = -1, length = entries == null ? 0 : entries.length;
			this.clear();
			while (++index < length) {
				var entry = entries[index];
				this.set(entry[0], entry[1]);
			}
		}
		ListCache.prototype.clear = listCacheClear;
		ListCache.prototype["delete"] = listCacheDelete;
		ListCache.prototype.get = listCacheGet;
		ListCache.prototype.has = listCacheHas;
		ListCache.prototype.set = listCacheSet;
		module.exports = ListCache;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_stackClear.js
	var require__stackClear = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var ListCache = require__ListCache();
		/**
		* Removes all key-value entries from the stack.
		*
		* @private
		* @name clear
		* @memberOf Stack
		*/
		function stackClear() {
			this.__data__ = new ListCache();
			this.size = 0;
		}
		module.exports = stackClear;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_stackDelete.js
	var require__stackDelete = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Removes `key` and its value from the stack.
		*
		* @private
		* @name delete
		* @memberOf Stack
		* @param {string} key The key of the value to remove.
		* @returns {boolean} Returns `true` if the entry was removed, else `false`.
		*/
		function stackDelete(key) {
			var data = this.__data__, result = data["delete"](key);
			this.size = data.size;
			return result;
		}
		module.exports = stackDelete;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_stackGet.js
	var require__stackGet = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Gets the stack value for `key`.
		*
		* @private
		* @name get
		* @memberOf Stack
		* @param {string} key The key of the value to get.
		* @returns {*} Returns the entry value.
		*/
		function stackGet(key) {
			return this.__data__.get(key);
		}
		module.exports = stackGet;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_stackHas.js
	var require__stackHas = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Checks if a stack value for `key` exists.
		*
		* @private
		* @name has
		* @memberOf Stack
		* @param {string} key The key of the entry to check.
		* @returns {boolean} Returns `true` if an entry for `key` exists, else `false`.
		*/
		function stackHas(key) {
			return this.__data__.has(key);
		}
		module.exports = stackHas;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isFunction.js
	var require_isFunction = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseGetTag = require__baseGetTag(), isObject = require_isObject();
		/** `Object#toString` result references. */
		var asyncTag = "[object AsyncFunction]", funcTag = "[object Function]", genTag = "[object GeneratorFunction]", proxyTag = "[object Proxy]";
		/**
		* Checks if `value` is classified as a `Function` object.
		*
		* @static
		* @memberOf _
		* @since 0.1.0
		* @category Lang
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is a function, else `false`.
		* @example
		*
		* _.isFunction(_);
		* // => true
		*
		* _.isFunction(/abc/);
		* // => false
		*/
		function isFunction(value) {
			if (!isObject(value)) return false;
			var tag = baseGetTag(value);
			return tag == funcTag || tag == genTag || tag == asyncTag || tag == proxyTag;
		}
		module.exports = isFunction;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_coreJsData.js
	var require__coreJsData = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require__root()["__core-js_shared__"];
	}));
	//#endregion
	//#region ../../node_modules/lodash/_isMasked.js
	var require__isMasked = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var coreJsData = require__coreJsData();
		/** Used to detect methods masquerading as native. */
		var maskSrcKey = function() {
			var uid = /[^.]+$/.exec(coreJsData && coreJsData.keys && coreJsData.keys.IE_PROTO || "");
			return uid ? "Symbol(src)_1." + uid : "";
		}();
		/**
		* Checks if `func` has its source masked.
		*
		* @private
		* @param {Function} func The function to check.
		* @returns {boolean} Returns `true` if `func` is masked, else `false`.
		*/
		function isMasked(func) {
			return !!maskSrcKey && maskSrcKey in func;
		}
		module.exports = isMasked;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_toSource.js
	var require__toSource = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/** Used to resolve the decompiled source of functions. */
		var funcToString = Function.prototype.toString;
		/**
		* Converts `func` to its source code.
		*
		* @private
		* @param {Function} func The function to convert.
		* @returns {string} Returns the source code.
		*/
		function toSource(func) {
			if (func != null) {
				try {
					return funcToString.call(func);
				} catch (e) {}
				try {
					return func + "";
				} catch (e) {}
			}
			return "";
		}
		module.exports = toSource;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_baseIsNative.js
	var require__baseIsNative = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var isFunction = require_isFunction(), isMasked = require__isMasked(), isObject = require_isObject(), toSource = require__toSource();
		/**
		* Used to match `RegExp`
		* [syntax characters](http://ecma-international.org/ecma-262/7.0/#sec-patterns).
		*/
		var reRegExpChar = /[\\^$.*+?()[\]{}|]/g;
		/** Used to detect host constructors (Safari). */
		var reIsHostCtor = /^\[object .+?Constructor\]$/;
		/** Used for built-in method references. */
		var funcProto = Function.prototype, objectProto = Object.prototype;
		/** Used to resolve the decompiled source of functions. */
		var funcToString = funcProto.toString;
		/** Used to check objects for own properties. */
		var hasOwnProperty = objectProto.hasOwnProperty;
		/** Used to detect if a method is native. */
		var reIsNative = RegExp("^" + funcToString.call(hasOwnProperty).replace(reRegExpChar, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$");
		/**
		* The base implementation of `_.isNative` without bad shim checks.
		*
		* @private
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is a native function,
		*  else `false`.
		*/
		function baseIsNative(value) {
			if (!isObject(value) || isMasked(value)) return false;
			return (isFunction(value) ? reIsNative : reIsHostCtor).test(toSource(value));
		}
		module.exports = baseIsNative;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_getValue.js
	var require__getValue = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Gets the value at `key` of `object`.
		*
		* @private
		* @param {Object} [object] The object to query.
		* @param {string} key The key of the property to get.
		* @returns {*} Returns the property value.
		*/
		function getValue(object, key) {
			return object == null ? void 0 : object[key];
		}
		module.exports = getValue;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_getNative.js
	var require__getNative = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseIsNative = require__baseIsNative(), getValue = require__getValue();
		/**
		* Gets the native function at `key` of `object`.
		*
		* @private
		* @param {Object} object The object to query.
		* @param {string} key The key of the method to get.
		* @returns {*} Returns the function if it's native, else `undefined`.
		*/
		function getNative(object, key) {
			var value = getValue(object, key);
			return baseIsNative(value) ? value : void 0;
		}
		module.exports = getNative;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_Map.js
	var require__Map = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require__getNative()(require__root(), "Map");
	}));
	//#endregion
	//#region ../../node_modules/lodash/_nativeCreate.js
	var require__nativeCreate = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require__getNative()(Object, "create");
	}));
	//#endregion
	//#region ../../node_modules/lodash/_hashClear.js
	var require__hashClear = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var nativeCreate = require__nativeCreate();
		/**
		* Removes all key-value entries from the hash.
		*
		* @private
		* @name clear
		* @memberOf Hash
		*/
		function hashClear() {
			this.__data__ = nativeCreate ? nativeCreate(null) : {};
			this.size = 0;
		}
		module.exports = hashClear;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_hashDelete.js
	var require__hashDelete = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Removes `key` and its value from the hash.
		*
		* @private
		* @name delete
		* @memberOf Hash
		* @param {Object} hash The hash to modify.
		* @param {string} key The key of the value to remove.
		* @returns {boolean} Returns `true` if the entry was removed, else `false`.
		*/
		function hashDelete(key) {
			var result = this.has(key) && delete this.__data__[key];
			this.size -= result ? 1 : 0;
			return result;
		}
		module.exports = hashDelete;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_hashGet.js
	var require__hashGet = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var nativeCreate = require__nativeCreate();
		/** Used to stand-in for `undefined` hash values. */
		var HASH_UNDEFINED = "__lodash_hash_undefined__";
		/** Used to check objects for own properties. */
		var hasOwnProperty = Object.prototype.hasOwnProperty;
		/**
		* Gets the hash value for `key`.
		*
		* @private
		* @name get
		* @memberOf Hash
		* @param {string} key The key of the value to get.
		* @returns {*} Returns the entry value.
		*/
		function hashGet(key) {
			var data = this.__data__;
			if (nativeCreate) {
				var result = data[key];
				return result === HASH_UNDEFINED ? void 0 : result;
			}
			return hasOwnProperty.call(data, key) ? data[key] : void 0;
		}
		module.exports = hashGet;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_hashHas.js
	var require__hashHas = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var nativeCreate = require__nativeCreate();
		/** Used to check objects for own properties. */
		var hasOwnProperty = Object.prototype.hasOwnProperty;
		/**
		* Checks if a hash value for `key` exists.
		*
		* @private
		* @name has
		* @memberOf Hash
		* @param {string} key The key of the entry to check.
		* @returns {boolean} Returns `true` if an entry for `key` exists, else `false`.
		*/
		function hashHas(key) {
			var data = this.__data__;
			return nativeCreate ? data[key] !== void 0 : hasOwnProperty.call(data, key);
		}
		module.exports = hashHas;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_hashSet.js
	var require__hashSet = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var nativeCreate = require__nativeCreate();
		/** Used to stand-in for `undefined` hash values. */
		var HASH_UNDEFINED = "__lodash_hash_undefined__";
		/**
		* Sets the hash `key` to `value`.
		*
		* @private
		* @name set
		* @memberOf Hash
		* @param {string} key The key of the value to set.
		* @param {*} value The value to set.
		* @returns {Object} Returns the hash instance.
		*/
		function hashSet(key, value) {
			var data = this.__data__;
			this.size += this.has(key) ? 0 : 1;
			data[key] = nativeCreate && value === void 0 ? HASH_UNDEFINED : value;
			return this;
		}
		module.exports = hashSet;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_Hash.js
	var require__Hash = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var hashClear = require__hashClear(), hashDelete = require__hashDelete(), hashGet = require__hashGet(), hashHas = require__hashHas(), hashSet = require__hashSet();
		/**
		* Creates a hash object.
		*
		* @private
		* @constructor
		* @param {Array} [entries] The key-value pairs to cache.
		*/
		function Hash(entries) {
			var index = -1, length = entries == null ? 0 : entries.length;
			this.clear();
			while (++index < length) {
				var entry = entries[index];
				this.set(entry[0], entry[1]);
			}
		}
		Hash.prototype.clear = hashClear;
		Hash.prototype["delete"] = hashDelete;
		Hash.prototype.get = hashGet;
		Hash.prototype.has = hashHas;
		Hash.prototype.set = hashSet;
		module.exports = Hash;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_mapCacheClear.js
	var require__mapCacheClear = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Hash = require__Hash(), ListCache = require__ListCache(), Map = require__Map();
		/**
		* Removes all key-value entries from the map.
		*
		* @private
		* @name clear
		* @memberOf MapCache
		*/
		function mapCacheClear() {
			this.size = 0;
			this.__data__ = {
				"hash": new Hash(),
				"map": new (Map || ListCache)(),
				"string": new Hash()
			};
		}
		module.exports = mapCacheClear;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_isKeyable.js
	var require__isKeyable = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Checks if `value` is suitable for use as unique object key.
		*
		* @private
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is suitable, else `false`.
		*/
		function isKeyable(value) {
			var type = typeof value;
			return type == "string" || type == "number" || type == "symbol" || type == "boolean" ? value !== "__proto__" : value === null;
		}
		module.exports = isKeyable;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_getMapData.js
	var require__getMapData = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var isKeyable = require__isKeyable();
		/**
		* Gets the data for `map`.
		*
		* @private
		* @param {Object} map The map to query.
		* @param {string} key The reference key.
		* @returns {*} Returns the map data.
		*/
		function getMapData(map, key) {
			var data = map.__data__;
			return isKeyable(key) ? data[typeof key == "string" ? "string" : "hash"] : data.map;
		}
		module.exports = getMapData;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_mapCacheDelete.js
	var require__mapCacheDelete = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var getMapData = require__getMapData();
		/**
		* Removes `key` and its value from the map.
		*
		* @private
		* @name delete
		* @memberOf MapCache
		* @param {string} key The key of the value to remove.
		* @returns {boolean} Returns `true` if the entry was removed, else `false`.
		*/
		function mapCacheDelete(key) {
			var result = getMapData(this, key)["delete"](key);
			this.size -= result ? 1 : 0;
			return result;
		}
		module.exports = mapCacheDelete;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_mapCacheGet.js
	var require__mapCacheGet = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var getMapData = require__getMapData();
		/**
		* Gets the map value for `key`.
		*
		* @private
		* @name get
		* @memberOf MapCache
		* @param {string} key The key of the value to get.
		* @returns {*} Returns the entry value.
		*/
		function mapCacheGet(key) {
			return getMapData(this, key).get(key);
		}
		module.exports = mapCacheGet;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_mapCacheHas.js
	var require__mapCacheHas = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var getMapData = require__getMapData();
		/**
		* Checks if a map value for `key` exists.
		*
		* @private
		* @name has
		* @memberOf MapCache
		* @param {string} key The key of the entry to check.
		* @returns {boolean} Returns `true` if an entry for `key` exists, else `false`.
		*/
		function mapCacheHas(key) {
			return getMapData(this, key).has(key);
		}
		module.exports = mapCacheHas;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_mapCacheSet.js
	var require__mapCacheSet = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var getMapData = require__getMapData();
		/**
		* Sets the map `key` to `value`.
		*
		* @private
		* @name set
		* @memberOf MapCache
		* @param {string} key The key of the value to set.
		* @param {*} value The value to set.
		* @returns {Object} Returns the map cache instance.
		*/
		function mapCacheSet(key, value) {
			var data = getMapData(this, key), size = data.size;
			data.set(key, value);
			this.size += data.size == size ? 0 : 1;
			return this;
		}
		module.exports = mapCacheSet;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_MapCache.js
	var require__MapCache = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var mapCacheClear = require__mapCacheClear(), mapCacheDelete = require__mapCacheDelete(), mapCacheGet = require__mapCacheGet(), mapCacheHas = require__mapCacheHas(), mapCacheSet = require__mapCacheSet();
		/**
		* Creates a map cache object to store key-value pairs.
		*
		* @private
		* @constructor
		* @param {Array} [entries] The key-value pairs to cache.
		*/
		function MapCache(entries) {
			var index = -1, length = entries == null ? 0 : entries.length;
			this.clear();
			while (++index < length) {
				var entry = entries[index];
				this.set(entry[0], entry[1]);
			}
		}
		MapCache.prototype.clear = mapCacheClear;
		MapCache.prototype["delete"] = mapCacheDelete;
		MapCache.prototype.get = mapCacheGet;
		MapCache.prototype.has = mapCacheHas;
		MapCache.prototype.set = mapCacheSet;
		module.exports = MapCache;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_stackSet.js
	var require__stackSet = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var ListCache = require__ListCache(), Map = require__Map(), MapCache = require__MapCache();
		/** Used as the size to enable large array optimizations. */
		var LARGE_ARRAY_SIZE = 200;
		/**
		* Sets the stack `key` to `value`.
		*
		* @private
		* @name set
		* @memberOf Stack
		* @param {string} key The key of the value to set.
		* @param {*} value The value to set.
		* @returns {Object} Returns the stack cache instance.
		*/
		function stackSet(key, value) {
			var data = this.__data__;
			if (data instanceof ListCache) {
				var pairs = data.__data__;
				if (!Map || pairs.length < LARGE_ARRAY_SIZE - 1) {
					pairs.push([key, value]);
					this.size = ++data.size;
					return this;
				}
				data = this.__data__ = new MapCache(pairs);
			}
			data.set(key, value);
			this.size = data.size;
			return this;
		}
		module.exports = stackSet;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_Stack.js
	var require__Stack = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var ListCache = require__ListCache(), stackClear = require__stackClear(), stackDelete = require__stackDelete(), stackGet = require__stackGet(), stackHas = require__stackHas(), stackSet = require__stackSet();
		/**
		* Creates a stack cache object to store key-value pairs.
		*
		* @private
		* @constructor
		* @param {Array} [entries] The key-value pairs to cache.
		*/
		function Stack(entries) {
			var data = this.__data__ = new ListCache(entries);
			this.size = data.size;
		}
		Stack.prototype.clear = stackClear;
		Stack.prototype["delete"] = stackDelete;
		Stack.prototype.get = stackGet;
		Stack.prototype.has = stackHas;
		Stack.prototype.set = stackSet;
		module.exports = Stack;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_setCacheAdd.js
	var require__setCacheAdd = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/** Used to stand-in for `undefined` hash values. */
		var HASH_UNDEFINED = "__lodash_hash_undefined__";
		/**
		* Adds `value` to the array cache.
		*
		* @private
		* @name add
		* @memberOf SetCache
		* @alias push
		* @param {*} value The value to cache.
		* @returns {Object} Returns the cache instance.
		*/
		function setCacheAdd(value) {
			this.__data__.set(value, HASH_UNDEFINED);
			return this;
		}
		module.exports = setCacheAdd;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_setCacheHas.js
	var require__setCacheHas = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Checks if `value` is in the array cache.
		*
		* @private
		* @name has
		* @memberOf SetCache
		* @param {*} value The value to search for.
		* @returns {boolean} Returns `true` if `value` is found, else `false`.
		*/
		function setCacheHas(value) {
			return this.__data__.has(value);
		}
		module.exports = setCacheHas;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_SetCache.js
	var require__SetCache = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var MapCache = require__MapCache(), setCacheAdd = require__setCacheAdd(), setCacheHas = require__setCacheHas();
		/**
		*
		* Creates an array cache object to store unique values.
		*
		* @private
		* @constructor
		* @param {Array} [values] The values to cache.
		*/
		function SetCache(values) {
			var index = -1, length = values == null ? 0 : values.length;
			this.__data__ = new MapCache();
			while (++index < length) this.add(values[index]);
		}
		SetCache.prototype.add = SetCache.prototype.push = setCacheAdd;
		SetCache.prototype.has = setCacheHas;
		module.exports = SetCache;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_arraySome.js
	var require__arraySome = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* A specialized version of `_.some` for arrays without support for iteratee
		* shorthands.
		*
		* @private
		* @param {Array} [array] The array to iterate over.
		* @param {Function} predicate The function invoked per iteration.
		* @returns {boolean} Returns `true` if any element passes the predicate check,
		*  else `false`.
		*/
		function arraySome(array, predicate) {
			var index = -1, length = array == null ? 0 : array.length;
			while (++index < length) if (predicate(array[index], index, array)) return true;
			return false;
		}
		module.exports = arraySome;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_cacheHas.js
	var require__cacheHas = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Checks if a `cache` value for `key` exists.
		*
		* @private
		* @param {Object} cache The cache to query.
		* @param {string} key The key of the entry to check.
		* @returns {boolean} Returns `true` if an entry for `key` exists, else `false`.
		*/
		function cacheHas(cache, key) {
			return cache.has(key);
		}
		module.exports = cacheHas;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_equalArrays.js
	var require__equalArrays = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var SetCache = require__SetCache(), arraySome = require__arraySome(), cacheHas = require__cacheHas();
		/** Used to compose bitmasks for value comparisons. */
		var COMPARE_PARTIAL_FLAG = 1, COMPARE_UNORDERED_FLAG = 2;
		/**
		* A specialized version of `baseIsEqualDeep` for arrays with support for
		* partial deep comparisons.
		*
		* @private
		* @param {Array} array The array to compare.
		* @param {Array} other The other array to compare.
		* @param {number} bitmask The bitmask flags. See `baseIsEqual` for more details.
		* @param {Function} customizer The function to customize comparisons.
		* @param {Function} equalFunc The function to determine equivalents of values.
		* @param {Object} stack Tracks traversed `array` and `other` objects.
		* @returns {boolean} Returns `true` if the arrays are equivalent, else `false`.
		*/
		function equalArrays(array, other, bitmask, customizer, equalFunc, stack) {
			var isPartial = bitmask & COMPARE_PARTIAL_FLAG, arrLength = array.length, othLength = other.length;
			if (arrLength != othLength && !(isPartial && othLength > arrLength)) return false;
			var arrStacked = stack.get(array);
			var othStacked = stack.get(other);
			if (arrStacked && othStacked) return arrStacked == other && othStacked == array;
			var index = -1, result = true, seen = bitmask & COMPARE_UNORDERED_FLAG ? new SetCache() : void 0;
			stack.set(array, other);
			stack.set(other, array);
			while (++index < arrLength) {
				var arrValue = array[index], othValue = other[index];
				if (customizer) var compared = isPartial ? customizer(othValue, arrValue, index, other, array, stack) : customizer(arrValue, othValue, index, array, other, stack);
				if (compared !== void 0) {
					if (compared) continue;
					result = false;
					break;
				}
				if (seen) {
					if (!arraySome(other, function(othValue, othIndex) {
						if (!cacheHas(seen, othIndex) && (arrValue === othValue || equalFunc(arrValue, othValue, bitmask, customizer, stack))) return seen.push(othIndex);
					})) {
						result = false;
						break;
					}
				} else if (!(arrValue === othValue || equalFunc(arrValue, othValue, bitmask, customizer, stack))) {
					result = false;
					break;
				}
			}
			stack["delete"](array);
			stack["delete"](other);
			return result;
		}
		module.exports = equalArrays;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_Uint8Array.js
	var require__Uint8Array = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require__root().Uint8Array;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_mapToArray.js
	var require__mapToArray = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Converts `map` to its key-value pairs.
		*
		* @private
		* @param {Object} map The map to convert.
		* @returns {Array} Returns the key-value pairs.
		*/
		function mapToArray(map) {
			var index = -1, result = Array(map.size);
			map.forEach(function(value, key) {
				result[++index] = [key, value];
			});
			return result;
		}
		module.exports = mapToArray;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_setToArray.js
	var require__setToArray = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Converts `set` to an array of its values.
		*
		* @private
		* @param {Object} set The set to convert.
		* @returns {Array} Returns the values.
		*/
		function setToArray(set) {
			var index = -1, result = Array(set.size);
			set.forEach(function(value) {
				result[++index] = value;
			});
			return result;
		}
		module.exports = setToArray;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_equalByTag.js
	var require__equalByTag = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Symbol = require__Symbol(), Uint8Array = require__Uint8Array(), eq = require_eq(), equalArrays = require__equalArrays(), mapToArray = require__mapToArray(), setToArray = require__setToArray();
		/** Used to compose bitmasks for value comparisons. */
		var COMPARE_PARTIAL_FLAG = 1, COMPARE_UNORDERED_FLAG = 2;
		/** `Object#toString` result references. */
		var boolTag = "[object Boolean]", dateTag = "[object Date]", errorTag = "[object Error]", mapTag = "[object Map]", numberTag = "[object Number]", regexpTag = "[object RegExp]", setTag = "[object Set]", stringTag = "[object String]", symbolTag = "[object Symbol]";
		var arrayBufferTag = "[object ArrayBuffer]", dataViewTag = "[object DataView]";
		/** Used to convert symbols to primitives and strings. */
		var symbolProto = Symbol ? Symbol.prototype : void 0, symbolValueOf = symbolProto ? symbolProto.valueOf : void 0;
		/**
		* A specialized version of `baseIsEqualDeep` for comparing objects of
		* the same `toStringTag`.
		*
		* **Note:** This function only supports comparing values with tags of
		* `Boolean`, `Date`, `Error`, `Number`, `RegExp`, or `String`.
		*
		* @private
		* @param {Object} object The object to compare.
		* @param {Object} other The other object to compare.
		* @param {string} tag The `toStringTag` of the objects to compare.
		* @param {number} bitmask The bitmask flags. See `baseIsEqual` for more details.
		* @param {Function} customizer The function to customize comparisons.
		* @param {Function} equalFunc The function to determine equivalents of values.
		* @param {Object} stack Tracks traversed `object` and `other` objects.
		* @returns {boolean} Returns `true` if the objects are equivalent, else `false`.
		*/
		function equalByTag(object, other, tag, bitmask, customizer, equalFunc, stack) {
			switch (tag) {
				case dataViewTag:
					if (object.byteLength != other.byteLength || object.byteOffset != other.byteOffset) return false;
					object = object.buffer;
					other = other.buffer;
				case arrayBufferTag:
					if (object.byteLength != other.byteLength || !equalFunc(new Uint8Array(object), new Uint8Array(other))) return false;
					return true;
				case boolTag:
				case dateTag:
				case numberTag: return eq(+object, +other);
				case errorTag: return object.name == other.name && object.message == other.message;
				case regexpTag:
				case stringTag: return object == other + "";
				case mapTag: var convert = mapToArray;
				case setTag:
					var isPartial = bitmask & COMPARE_PARTIAL_FLAG;
					convert || (convert = setToArray);
					if (object.size != other.size && !isPartial) return false;
					var stacked = stack.get(object);
					if (stacked) return stacked == other;
					bitmask |= COMPARE_UNORDERED_FLAG;
					stack.set(object, other);
					var result = equalArrays(convert(object), convert(other), bitmask, customizer, equalFunc, stack);
					stack["delete"](object);
					return result;
				case symbolTag: if (symbolValueOf) return symbolValueOf.call(object) == symbolValueOf.call(other);
			}
			return false;
		}
		module.exports = equalByTag;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_arrayPush.js
	var require__arrayPush = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Appends the elements of `values` to `array`.
		*
		* @private
		* @param {Array} array The array to modify.
		* @param {Array} values The values to append.
		* @returns {Array} Returns `array`.
		*/
		function arrayPush(array, values) {
			var index = -1, length = values.length, offset = array.length;
			while (++index < length) array[offset + index] = values[index];
			return array;
		}
		module.exports = arrayPush;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_baseGetAllKeys.js
	var require__baseGetAllKeys = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var arrayPush = require__arrayPush(), isArray = require_isArray();
		/**
		* The base implementation of `getAllKeys` and `getAllKeysIn` which uses
		* `keysFunc` and `symbolsFunc` to get the enumerable property names and
		* symbols of `object`.
		*
		* @private
		* @param {Object} object The object to query.
		* @param {Function} keysFunc The function to get the keys of `object`.
		* @param {Function} symbolsFunc The function to get the symbols of `object`.
		* @returns {Array} Returns the array of property names and symbols.
		*/
		function baseGetAllKeys(object, keysFunc, symbolsFunc) {
			var result = keysFunc(object);
			return isArray(object) ? result : arrayPush(result, symbolsFunc(object));
		}
		module.exports = baseGetAllKeys;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_arrayFilter.js
	var require__arrayFilter = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* A specialized version of `_.filter` for arrays without support for
		* iteratee shorthands.
		*
		* @private
		* @param {Array} [array] The array to iterate over.
		* @param {Function} predicate The function invoked per iteration.
		* @returns {Array} Returns the new filtered array.
		*/
		function arrayFilter(array, predicate) {
			var index = -1, length = array == null ? 0 : array.length, resIndex = 0, result = [];
			while (++index < length) {
				var value = array[index];
				if (predicate(value, index, array)) result[resIndex++] = value;
			}
			return result;
		}
		module.exports = arrayFilter;
	}));
	//#endregion
	//#region ../../node_modules/lodash/stubArray.js
	var require_stubArray = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* This method returns a new empty array.
		*
		* @static
		* @memberOf _
		* @since 4.13.0
		* @category Util
		* @returns {Array} Returns the new empty array.
		* @example
		*
		* var arrays = _.times(2, _.stubArray);
		*
		* console.log(arrays);
		* // => [[], []]
		*
		* console.log(arrays[0] === arrays[1]);
		* // => false
		*/
		function stubArray() {
			return [];
		}
		module.exports = stubArray;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_getSymbols.js
	var require__getSymbols = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var arrayFilter = require__arrayFilter(), stubArray = require_stubArray();
		/** Built-in value references. */
		var propertyIsEnumerable = Object.prototype.propertyIsEnumerable;
		var nativeGetSymbols = Object.getOwnPropertySymbols;
		module.exports = !nativeGetSymbols ? stubArray : function(object) {
			if (object == null) return [];
			object = Object(object);
			return arrayFilter(nativeGetSymbols(object), function(symbol) {
				return propertyIsEnumerable.call(object, symbol);
			});
		};
	}));
	//#endregion
	//#region ../../node_modules/lodash/_baseTimes.js
	var require__baseTimes = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* The base implementation of `_.times` without support for iteratee shorthands
		* or max array length checks.
		*
		* @private
		* @param {number} n The number of times to invoke `iteratee`.
		* @param {Function} iteratee The function invoked per iteration.
		* @returns {Array} Returns the array of results.
		*/
		function baseTimes(n, iteratee) {
			var index = -1, result = Array(n);
			while (++index < n) result[index] = iteratee(index);
			return result;
		}
		module.exports = baseTimes;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_baseIsArguments.js
	var require__baseIsArguments = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseGetTag = require__baseGetTag(), isObjectLike = require_isObjectLike();
		/** `Object#toString` result references. */
		var argsTag = "[object Arguments]";
		/**
		* The base implementation of `_.isArguments`.
		*
		* @private
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is an `arguments` object,
		*/
		function baseIsArguments(value) {
			return isObjectLike(value) && baseGetTag(value) == argsTag;
		}
		module.exports = baseIsArguments;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isArguments.js
	var require_isArguments = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseIsArguments = require__baseIsArguments(), isObjectLike = require_isObjectLike();
		/** Used for built-in method references. */
		var objectProto = Object.prototype;
		/** Used to check objects for own properties. */
		var hasOwnProperty = objectProto.hasOwnProperty;
		/** Built-in value references. */
		var propertyIsEnumerable = objectProto.propertyIsEnumerable;
		module.exports = baseIsArguments(function() {
			return arguments;
		}()) ? baseIsArguments : function(value) {
			return isObjectLike(value) && hasOwnProperty.call(value, "callee") && !propertyIsEnumerable.call(value, "callee");
		};
	}));
	//#endregion
	//#region ../../node_modules/lodash/stubFalse.js
	var require_stubFalse = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* This method returns `false`.
		*
		* @static
		* @memberOf _
		* @since 4.13.0
		* @category Util
		* @returns {boolean} Returns `false`.
		* @example
		*
		* _.times(2, _.stubFalse);
		* // => [false, false]
		*/
		function stubFalse() {
			return false;
		}
		module.exports = stubFalse;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isBuffer.js
	var require_isBuffer = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var root = require__root(), stubFalse = require_stubFalse();
		/** Detect free variable `exports`. */
		var freeExports = typeof exports == "object" && exports && !exports.nodeType && exports;
		/** Detect free variable `module`. */
		var freeModule = freeExports && typeof module == "object" && module && !module.nodeType && module;
		/** Built-in value references. */
		var Buffer = freeModule && freeModule.exports === freeExports ? root.Buffer : void 0;
		module.exports = (Buffer ? Buffer.isBuffer : void 0) || stubFalse;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_isIndex.js
	var require__isIndex = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/** Used as references for various `Number` constants. */
		var MAX_SAFE_INTEGER = 9007199254740991;
		/** Used to detect unsigned integer values. */
		var reIsUint = /^(?:0|[1-9]\d*)$/;
		/**
		* Checks if `value` is a valid array-like index.
		*
		* @private
		* @param {*} value The value to check.
		* @param {number} [length=MAX_SAFE_INTEGER] The upper bounds of a valid index.
		* @returns {boolean} Returns `true` if `value` is a valid index, else `false`.
		*/
		function isIndex(value, length) {
			var type = typeof value;
			length = length == null ? MAX_SAFE_INTEGER : length;
			return !!length && (type == "number" || type != "symbol" && reIsUint.test(value)) && value > -1 && value % 1 == 0 && value < length;
		}
		module.exports = isIndex;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isLength.js
	var require_isLength = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/** Used as references for various `Number` constants. */
		var MAX_SAFE_INTEGER = 9007199254740991;
		/**
		* Checks if `value` is a valid array-like length.
		*
		* **Note:** This method is loosely based on
		* [`ToLength`](http://ecma-international.org/ecma-262/7.0/#sec-tolength).
		*
		* @static
		* @memberOf _
		* @since 4.0.0
		* @category Lang
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is a valid length, else `false`.
		* @example
		*
		* _.isLength(3);
		* // => true
		*
		* _.isLength(Number.MIN_VALUE);
		* // => false
		*
		* _.isLength(Infinity);
		* // => false
		*
		* _.isLength('3');
		* // => false
		*/
		function isLength(value) {
			return typeof value == "number" && value > -1 && value % 1 == 0 && value <= MAX_SAFE_INTEGER;
		}
		module.exports = isLength;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_baseIsTypedArray.js
	var require__baseIsTypedArray = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseGetTag = require__baseGetTag(), isLength = require_isLength(), isObjectLike = require_isObjectLike();
		/** `Object#toString` result references. */
		var argsTag = "[object Arguments]", arrayTag = "[object Array]", boolTag = "[object Boolean]", dateTag = "[object Date]", errorTag = "[object Error]", funcTag = "[object Function]", mapTag = "[object Map]", numberTag = "[object Number]", objectTag = "[object Object]", regexpTag = "[object RegExp]", setTag = "[object Set]", stringTag = "[object String]", weakMapTag = "[object WeakMap]";
		var arrayBufferTag = "[object ArrayBuffer]", dataViewTag = "[object DataView]", float32Tag = "[object Float32Array]", float64Tag = "[object Float64Array]", int8Tag = "[object Int8Array]", int16Tag = "[object Int16Array]", int32Tag = "[object Int32Array]", uint8Tag = "[object Uint8Array]", uint8ClampedTag = "[object Uint8ClampedArray]", uint16Tag = "[object Uint16Array]", uint32Tag = "[object Uint32Array]";
		/** Used to identify `toStringTag` values of typed arrays. */
		var typedArrayTags = {};
		typedArrayTags[float32Tag] = typedArrayTags[float64Tag] = typedArrayTags[int8Tag] = typedArrayTags[int16Tag] = typedArrayTags[int32Tag] = typedArrayTags[uint8Tag] = typedArrayTags[uint8ClampedTag] = typedArrayTags[uint16Tag] = typedArrayTags[uint32Tag] = true;
		typedArrayTags[argsTag] = typedArrayTags[arrayTag] = typedArrayTags[arrayBufferTag] = typedArrayTags[boolTag] = typedArrayTags[dataViewTag] = typedArrayTags[dateTag] = typedArrayTags[errorTag] = typedArrayTags[funcTag] = typedArrayTags[mapTag] = typedArrayTags[numberTag] = typedArrayTags[objectTag] = typedArrayTags[regexpTag] = typedArrayTags[setTag] = typedArrayTags[stringTag] = typedArrayTags[weakMapTag] = false;
		/**
		* The base implementation of `_.isTypedArray` without Node.js optimizations.
		*
		* @private
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is a typed array, else `false`.
		*/
		function baseIsTypedArray(value) {
			return isObjectLike(value) && isLength(value.length) && !!typedArrayTags[baseGetTag(value)];
		}
		module.exports = baseIsTypedArray;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_baseUnary.js
	var require__baseUnary = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* The base implementation of `_.unary` without support for storing metadata.
		*
		* @private
		* @param {Function} func The function to cap arguments for.
		* @returns {Function} Returns the new capped function.
		*/
		function baseUnary(func) {
			return function(value) {
				return func(value);
			};
		}
		module.exports = baseUnary;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_nodeUtil.js
	var require__nodeUtil = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var freeGlobal = require__freeGlobal();
		/** Detect free variable `exports`. */
		var freeExports = typeof exports == "object" && exports && !exports.nodeType && exports;
		/** Detect free variable `module`. */
		var freeModule = freeExports && typeof module == "object" && module && !module.nodeType && module;
		/** Detect free variable `process` from Node.js. */
		var freeProcess = freeModule && freeModule.exports === freeExports && freeGlobal.process;
		module.exports = function() {
			try {
				var types = freeModule && freeModule.require && freeModule.require("util").types;
				if (types) return types;
				return freeProcess && freeProcess.binding && freeProcess.binding("util");
			} catch (e) {}
		}();
	}));
	//#endregion
	//#region ../../node_modules/lodash/isTypedArray.js
	var require_isTypedArray = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseIsTypedArray = require__baseIsTypedArray(), baseUnary = require__baseUnary(), nodeUtil = require__nodeUtil();
		var nodeIsTypedArray = nodeUtil && nodeUtil.isTypedArray;
		module.exports = nodeIsTypedArray ? baseUnary(nodeIsTypedArray) : baseIsTypedArray;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_arrayLikeKeys.js
	var require__arrayLikeKeys = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseTimes = require__baseTimes(), isArguments = require_isArguments(), isArray = require_isArray(), isBuffer = require_isBuffer(), isIndex = require__isIndex(), isTypedArray = require_isTypedArray();
		/** Used to check objects for own properties. */
		var hasOwnProperty = Object.prototype.hasOwnProperty;
		/**
		* Creates an array of the enumerable property names of the array-like `value`.
		*
		* @private
		* @param {*} value The value to query.
		* @param {boolean} inherited Specify returning inherited property names.
		* @returns {Array} Returns the array of property names.
		*/
		function arrayLikeKeys(value, inherited) {
			var isArr = isArray(value), isArg = !isArr && isArguments(value), isBuff = !isArr && !isArg && isBuffer(value), isType = !isArr && !isArg && !isBuff && isTypedArray(value), skipIndexes = isArr || isArg || isBuff || isType, result = skipIndexes ? baseTimes(value.length, String) : [], length = result.length;
			for (var key in value) if ((inherited || hasOwnProperty.call(value, key)) && !(skipIndexes && (key == "length" || isBuff && (key == "offset" || key == "parent") || isType && (key == "buffer" || key == "byteLength" || key == "byteOffset") || isIndex(key, length)))) result.push(key);
			return result;
		}
		module.exports = arrayLikeKeys;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_isPrototype.js
	var require__isPrototype = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/** Used for built-in method references. */
		var objectProto = Object.prototype;
		/**
		* Checks if `value` is likely a prototype object.
		*
		* @private
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is a prototype, else `false`.
		*/
		function isPrototype(value) {
			var Ctor = value && value.constructor;
			return value === (typeof Ctor == "function" && Ctor.prototype || objectProto);
		}
		module.exports = isPrototype;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_overArg.js
	var require__overArg = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/**
		* Creates a unary function that invokes `func` with its argument transformed.
		*
		* @private
		* @param {Function} func The function to wrap.
		* @param {Function} transform The argument transform.
		* @returns {Function} Returns the new function.
		*/
		function overArg(func, transform) {
			return function(arg) {
				return func(transform(arg));
			};
		}
		module.exports = overArg;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_nativeKeys.js
	var require__nativeKeys = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require__overArg()(Object.keys, Object);
	}));
	//#endregion
	//#region ../../node_modules/lodash/_baseKeys.js
	var require__baseKeys = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var isPrototype = require__isPrototype(), nativeKeys = require__nativeKeys();
		/** Used to check objects for own properties. */
		var hasOwnProperty = Object.prototype.hasOwnProperty;
		/**
		* The base implementation of `_.keys` which doesn't treat sparse arrays as dense.
		*
		* @private
		* @param {Object} object The object to query.
		* @returns {Array} Returns the array of property names.
		*/
		function baseKeys(object) {
			if (!isPrototype(object)) return nativeKeys(object);
			var result = [];
			for (var key in Object(object)) if (hasOwnProperty.call(object, key) && key != "constructor") result.push(key);
			return result;
		}
		module.exports = baseKeys;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isArrayLike.js
	var require_isArrayLike = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var isFunction = require_isFunction(), isLength = require_isLength();
		/**
		* Checks if `value` is array-like. A value is considered array-like if it's
		* not a function and has a `value.length` that's an integer greater than or
		* equal to `0` and less than or equal to `Number.MAX_SAFE_INTEGER`.
		*
		* @static
		* @memberOf _
		* @since 4.0.0
		* @category Lang
		* @param {*} value The value to check.
		* @returns {boolean} Returns `true` if `value` is array-like, else `false`.
		* @example
		*
		* _.isArrayLike([1, 2, 3]);
		* // => true
		*
		* _.isArrayLike(document.body.children);
		* // => true
		*
		* _.isArrayLike('abc');
		* // => true
		*
		* _.isArrayLike(_.noop);
		* // => false
		*/
		function isArrayLike(value) {
			return value != null && isLength(value.length) && !isFunction(value);
		}
		module.exports = isArrayLike;
	}));
	//#endregion
	//#region ../../node_modules/lodash/keys.js
	var require_keys = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var arrayLikeKeys = require__arrayLikeKeys(), baseKeys = require__baseKeys(), isArrayLike = require_isArrayLike();
		/**
		* Creates an array of the own enumerable property names of `object`.
		*
		* **Note:** Non-object values are coerced to objects. See the
		* [ES spec](http://ecma-international.org/ecma-262/7.0/#sec-object.keys)
		* for more details.
		*
		* @static
		* @since 0.1.0
		* @memberOf _
		* @category Object
		* @param {Object} object The object to query.
		* @returns {Array} Returns the array of property names.
		* @example
		*
		* function Foo() {
		*   this.a = 1;
		*   this.b = 2;
		* }
		*
		* Foo.prototype.c = 3;
		*
		* _.keys(new Foo);
		* // => ['a', 'b'] (iteration order is not guaranteed)
		*
		* _.keys('hi');
		* // => ['0', '1']
		*/
		function keys(object) {
			return isArrayLike(object) ? arrayLikeKeys(object) : baseKeys(object);
		}
		module.exports = keys;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_getAllKeys.js
	var require__getAllKeys = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseGetAllKeys = require__baseGetAllKeys(), getSymbols = require__getSymbols(), keys = require_keys();
		/**
		* Creates an array of own enumerable property names and symbols of `object`.
		*
		* @private
		* @param {Object} object The object to query.
		* @returns {Array} Returns the array of property names and symbols.
		*/
		function getAllKeys(object) {
			return baseGetAllKeys(object, keys, getSymbols);
		}
		module.exports = getAllKeys;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_equalObjects.js
	var require__equalObjects = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var getAllKeys = require__getAllKeys();
		/** Used to compose bitmasks for value comparisons. */
		var COMPARE_PARTIAL_FLAG = 1;
		/** Used to check objects for own properties. */
		var hasOwnProperty = Object.prototype.hasOwnProperty;
		/**
		* A specialized version of `baseIsEqualDeep` for objects with support for
		* partial deep comparisons.
		*
		* @private
		* @param {Object} object The object to compare.
		* @param {Object} other The other object to compare.
		* @param {number} bitmask The bitmask flags. See `baseIsEqual` for more details.
		* @param {Function} customizer The function to customize comparisons.
		* @param {Function} equalFunc The function to determine equivalents of values.
		* @param {Object} stack Tracks traversed `object` and `other` objects.
		* @returns {boolean} Returns `true` if the objects are equivalent, else `false`.
		*/
		function equalObjects(object, other, bitmask, customizer, equalFunc, stack) {
			var isPartial = bitmask & COMPARE_PARTIAL_FLAG, objProps = getAllKeys(object), objLength = objProps.length;
			if (objLength != getAllKeys(other).length && !isPartial) return false;
			var index = objLength;
			while (index--) {
				var key = objProps[index];
				if (!(isPartial ? key in other : hasOwnProperty.call(other, key))) return false;
			}
			var objStacked = stack.get(object);
			var othStacked = stack.get(other);
			if (objStacked && othStacked) return objStacked == other && othStacked == object;
			var result = true;
			stack.set(object, other);
			stack.set(other, object);
			var skipCtor = isPartial;
			while (++index < objLength) {
				key = objProps[index];
				var objValue = object[key], othValue = other[key];
				if (customizer) var compared = isPartial ? customizer(othValue, objValue, key, other, object, stack) : customizer(objValue, othValue, key, object, other, stack);
				if (!(compared === void 0 ? objValue === othValue || equalFunc(objValue, othValue, bitmask, customizer, stack) : compared)) {
					result = false;
					break;
				}
				skipCtor || (skipCtor = key == "constructor");
			}
			if (result && !skipCtor) {
				var objCtor = object.constructor, othCtor = other.constructor;
				if (objCtor != othCtor && "constructor" in object && "constructor" in other && !(typeof objCtor == "function" && objCtor instanceof objCtor && typeof othCtor == "function" && othCtor instanceof othCtor)) result = false;
			}
			stack["delete"](object);
			stack["delete"](other);
			return result;
		}
		module.exports = equalObjects;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_DataView.js
	var require__DataView = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require__getNative()(require__root(), "DataView");
	}));
	//#endregion
	//#region ../../node_modules/lodash/_Promise.js
	var require__Promise = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require__getNative()(require__root(), "Promise");
	}));
	//#endregion
	//#region ../../node_modules/lodash/_Set.js
	var require__Set = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require__getNative()(require__root(), "Set");
	}));
	//#endregion
	//#region ../../node_modules/lodash/_WeakMap.js
	var require__WeakMap = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = require__getNative()(require__root(), "WeakMap");
	}));
	//#endregion
	//#region ../../node_modules/lodash/_getTag.js
	var require__getTag = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var DataView = require__DataView(), Map = require__Map(), Promise = require__Promise(), Set = require__Set(), WeakMap = require__WeakMap(), baseGetTag = require__baseGetTag(), toSource = require__toSource();
		/** `Object#toString` result references. */
		var mapTag = "[object Map]", objectTag = "[object Object]", promiseTag = "[object Promise]", setTag = "[object Set]", weakMapTag = "[object WeakMap]";
		var dataViewTag = "[object DataView]";
		/** Used to detect maps, sets, and weakmaps. */
		var dataViewCtorString = toSource(DataView), mapCtorString = toSource(Map), promiseCtorString = toSource(Promise), setCtorString = toSource(Set), weakMapCtorString = toSource(WeakMap);
		/**
		* Gets the `toStringTag` of `value`.
		*
		* @private
		* @param {*} value The value to query.
		* @returns {string} Returns the `toStringTag`.
		*/
		var getTag = baseGetTag;
		if (DataView && getTag(new DataView(/* @__PURE__ */ new ArrayBuffer(1))) != dataViewTag || Map && getTag(new Map()) != mapTag || Promise && getTag(Promise.resolve()) != promiseTag || Set && getTag(new Set()) != setTag || WeakMap && getTag(new WeakMap()) != weakMapTag) getTag = function(value) {
			var result = baseGetTag(value), Ctor = result == objectTag ? value.constructor : void 0, ctorString = Ctor ? toSource(Ctor) : "";
			if (ctorString) switch (ctorString) {
				case dataViewCtorString: return dataViewTag;
				case mapCtorString: return mapTag;
				case promiseCtorString: return promiseTag;
				case setCtorString: return setTag;
				case weakMapCtorString: return weakMapTag;
			}
			return result;
		};
		module.exports = getTag;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_baseIsEqualDeep.js
	var require__baseIsEqualDeep = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Stack = require__Stack(), equalArrays = require__equalArrays(), equalByTag = require__equalByTag(), equalObjects = require__equalObjects(), getTag = require__getTag(), isArray = require_isArray(), isBuffer = require_isBuffer(), isTypedArray = require_isTypedArray();
		/** Used to compose bitmasks for value comparisons. */
		var COMPARE_PARTIAL_FLAG = 1;
		/** `Object#toString` result references. */
		var argsTag = "[object Arguments]", arrayTag = "[object Array]", objectTag = "[object Object]";
		/** Used to check objects for own properties. */
		var hasOwnProperty = Object.prototype.hasOwnProperty;
		/**
		* A specialized version of `baseIsEqual` for arrays and objects which performs
		* deep comparisons and tracks traversed objects enabling objects with circular
		* references to be compared.
		*
		* @private
		* @param {Object} object The object to compare.
		* @param {Object} other The other object to compare.
		* @param {number} bitmask The bitmask flags. See `baseIsEqual` for more details.
		* @param {Function} customizer The function to customize comparisons.
		* @param {Function} equalFunc The function to determine equivalents of values.
		* @param {Object} [stack] Tracks traversed `object` and `other` objects.
		* @returns {boolean} Returns `true` if the objects are equivalent, else `false`.
		*/
		function baseIsEqualDeep(object, other, bitmask, customizer, equalFunc, stack) {
			var objIsArr = isArray(object), othIsArr = isArray(other), objTag = objIsArr ? arrayTag : getTag(object), othTag = othIsArr ? arrayTag : getTag(other);
			objTag = objTag == argsTag ? objectTag : objTag;
			othTag = othTag == argsTag ? objectTag : othTag;
			var objIsObj = objTag == objectTag, othIsObj = othTag == objectTag, isSameTag = objTag == othTag;
			if (isSameTag && isBuffer(object)) {
				if (!isBuffer(other)) return false;
				objIsArr = true;
				objIsObj = false;
			}
			if (isSameTag && !objIsObj) {
				stack || (stack = new Stack());
				return objIsArr || isTypedArray(object) ? equalArrays(object, other, bitmask, customizer, equalFunc, stack) : equalByTag(object, other, objTag, bitmask, customizer, equalFunc, stack);
			}
			if (!(bitmask & COMPARE_PARTIAL_FLAG)) {
				var objIsWrapped = objIsObj && hasOwnProperty.call(object, "__wrapped__"), othIsWrapped = othIsObj && hasOwnProperty.call(other, "__wrapped__");
				if (objIsWrapped || othIsWrapped) {
					var objUnwrapped = objIsWrapped ? object.value() : object, othUnwrapped = othIsWrapped ? other.value() : other;
					stack || (stack = new Stack());
					return equalFunc(objUnwrapped, othUnwrapped, bitmask, customizer, stack);
				}
			}
			if (!isSameTag) return false;
			stack || (stack = new Stack());
			return equalObjects(object, other, bitmask, customizer, equalFunc, stack);
		}
		module.exports = baseIsEqualDeep;
	}));
	//#endregion
	//#region ../../node_modules/lodash/_baseIsEqual.js
	var require__baseIsEqual = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseIsEqualDeep = require__baseIsEqualDeep(), isObjectLike = require_isObjectLike();
		/**
		* The base implementation of `_.isEqual` which supports partial comparisons
		* and tracks traversed objects.
		*
		* @private
		* @param {*} value The value to compare.
		* @param {*} other The other value to compare.
		* @param {boolean} bitmask The bitmask flags.
		*  1 - Unordered comparison
		*  2 - Partial comparison
		* @param {Function} [customizer] The function to customize comparisons.
		* @param {Object} [stack] Tracks traversed `value` and `other` objects.
		* @returns {boolean} Returns `true` if the values are equivalent, else `false`.
		*/
		function baseIsEqual(value, other, bitmask, customizer, stack) {
			if (value === other) return true;
			if (value == null || other == null || !isObjectLike(value) && !isObjectLike(other)) return value !== value && other !== other;
			return baseIsEqualDeep(value, other, bitmask, customizer, baseIsEqual, stack);
		}
		module.exports = baseIsEqual;
	}));
	//#endregion
	//#region ../../node_modules/lodash/isEqual.js
	var require_isEqual = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var baseIsEqual = require__baseIsEqual();
		/**
		* Performs a deep comparison between two values to determine if they are
		* equivalent.
		*
		* **Note:** This method supports comparing arrays, array buffers, booleans,
		* date objects, error objects, maps, numbers, `Object` objects, regexes,
		* sets, strings, symbols, and typed arrays. `Object` objects are compared
		* by their own, not inherited, enumerable properties. Functions and DOM
		* nodes are compared by strict equality, i.e. `===`.
		*
		* @static
		* @memberOf _
		* @since 0.1.0
		* @category Lang
		* @param {*} value The value to compare.
		* @param {*} other The other value to compare.
		* @returns {boolean} Returns `true` if the values are equivalent, else `false`.
		* @example
		*
		* var object = { 'a': 1 };
		* var other = { 'a': 1 };
		*
		* _.isEqual(object, other);
		* // => true
		*
		* object === other;
		* // => false
		*/
		function isEqual(value, other) {
			return baseIsEqual(value, other);
		}
		module.exports = isEqual;
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/KeyValuePair.js
	var require_KeyValuePair = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		module.exports = class KeyValuePair {
			constructor(key, value) {
				this.key = key;
				this.value = value;
			}
			/**
			* @returns {KeyValuePair}
			*/
			clone() {
				const clone = new KeyValuePair();
				if (this.key) clone.key = this.key.clone();
				if (this.value) clone.value = this.value.clone();
				return clone;
			}
		};
	}));
	//#endregion
	//#region ../../node_modules/lodash/negate.js
	var require_negate = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		/** Error message constants. */
		var FUNC_ERROR_TEXT = "Expected a function";
		/**
		* Creates a function that negates the result of the predicate `func`. The
		* `func` predicate is invoked with the `this` binding and arguments of the
		* created function.
		*
		* @static
		* @memberOf _
		* @since 3.0.0
		* @category Function
		* @param {Function} predicate The predicate to negate.
		* @returns {Function} Returns the new negated function.
		* @example
		*
		* function isEven(n) {
		*   return n % 2 == 0;
		* }
		*
		* _.filter([1, 2, 3, 4, 5, 6], _.negate(isEven));
		* // => [1, 3, 5]
		*/
		function negate(predicate) {
			if (typeof predicate != "function") throw new TypeError(FUNC_ERROR_TEXT);
			return function() {
				var args = arguments;
				switch (args.length) {
					case 0: return !predicate.call(this);
					case 1: return !predicate.call(this, args[0]);
					case 2: return !predicate.call(this, args[0], args[1]);
					case 3: return !predicate.call(this, args[0], args[1], args[2]);
				}
				return !predicate.apply(this, args);
			};
		}
		module.exports = negate;
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/ArraySlice.js
	var require_ArraySlice = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var negate = require_negate();
		function coerceElementMatchingCallback(value) {
			if (typeof value === "string") return (element) => element.element === value;
			if (value.constructor && value.extend) return (element) => element instanceof value;
			return value;
		}
		/**
		* @class
		*
		* @param {Element[]} elements
		*
		* @property {Element[]} elements
		*/
		var ArraySlice = class ArraySlice {
			constructor(elements) {
				this.elements = elements || [];
			}
			/**
			* @returns {Array}
			*/
			toValue() {
				return this.elements.map((element) => element.toValue());
			}
			/**
			* @param callback - Function to execute for each element
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @returns {array} A new array with each element being the result of the callback function
			*/
			map(callback, thisArg) {
				return this.elements.map(callback, thisArg);
			}
			/**
			* Maps and then flattens the results.
			* @param callback - Function to execute for each element.
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @returns {array}
			*/
			flatMap(callback, thisArg) {
				return this.map(callback, thisArg).reduce((a, b) => a.concat(b), []);
			}
			/**
			* Returns an array containing the truthy results of calling the given transformation with each element of this sequence
			* @param transform - A closure that accepts an element of this array as its argument and returns an optional value.
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @memberof ArraySlice.prototype
			* @returns An array of the non-undefined results of calling transform with each element of the array
			*/
			compactMap(transform, thisArg) {
				const results = [];
				this.forEach((element) => {
					const result = transform.bind(thisArg)(element);
					if (result) results.push(result);
				});
				return results;
			}
			/**
			* @param callback - Function to execute for each element. This may be a callback, an element name or an element class.
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @returns {ArraySlice}
			* @memberof ArraySlice.prototype
			*/
			filter(callback, thisArg) {
				callback = coerceElementMatchingCallback(callback);
				return new ArraySlice(this.elements.filter(callback, thisArg));
			}
			/**
			* @param callback - Function to execute for each element. This may be a callback, an element name or an element class.
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @returns {ArraySlice}
			* @memberof ArraySlice.prototype
			*/
			reject(callback, thisArg) {
				callback = coerceElementMatchingCallback(callback);
				return new ArraySlice(this.elements.filter(negate(callback), thisArg));
			}
			/**
			* Returns the first element in the array that satisfies the given value
			* @param callback - Function to execute for each element. This may be a callback, an element name or an element class.
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @returns {Element}
			* @memberof ArraySlice.prototype
			*/
			find(callback, thisArg) {
				callback = coerceElementMatchingCallback(callback);
				return this.elements.find(callback, thisArg);
			}
			/**
			* @param callback - Function to execute for each element
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @memberof ArraySlice.prototype
			*/
			forEach(callback, thisArg) {
				this.elements.forEach(callback, thisArg);
			}
			/**
			* @param callback - Function to execute for each element
			* @param initialValue
			* @memberof ArraySlice.prototype
			*/
			reduce(callback, initialValue) {
				return this.elements.reduce(callback, initialValue);
			}
			/**
			* @param value
			* @returns {boolean}
			* @memberof ArraySlice.prototype
			*/
			includes(value) {
				return this.elements.some((element) => element.equals(value));
			}
			/**
			* Removes the first element from the slice
			* @returns {Element} The removed element or undefined if the slice is empty
			* @memberof ArraySlice.prototype
			*/
			shift() {
				return this.elements.shift();
			}
			/**
			* Adds the given element to the begining of the slice
			* @parameter {Element} value
			* @memberof ArraySlice.prototype
			*/
			unshift(value) {
				this.elements.unshift(this.refract(value));
			}
			/**
			* Adds the given element to the end of the slice
			* @parameter {Element} value
			* @memberof ArraySlice.prototype
			*/
			push(value) {
				this.elements.push(this.refract(value));
				return this;
			}
			/**
			* @parameter {Element} value
			* @memberof ArraySlice.prototype
			*/
			add(value) {
				this.push(value);
			}
			/**
			* @parameter {number} index
			* @returns {Element}
			* @memberof ArraySlice.prototype
			*/
			get(index) {
				return this.elements[index];
			}
			/**
			* @parameter {number} index
			* @memberof ArraySlice.prototype
			*/
			getValue(index) {
				const element = this.elements[index];
				if (element) return element.toValue();
			}
			/**
			* Returns the number of elements in the slice
			* @type number
			*/
			get length() {
				return this.elements.length;
			}
			/**
			* Returns whether the slice is empty
			* @type boolean
			*/
			get isEmpty() {
				return this.elements.length === 0;
			}
			/**
			* Returns the first element in the slice or undefined if the slice is empty
			* @type Element
			*/
			get first() {
				return this.elements[0];
			}
		};
		if (typeof Symbol !== "undefined") ArraySlice.prototype[Symbol.iterator] = function symbol() {
			return this.elements[Symbol.iterator]();
		};
		module.exports = ArraySlice;
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/primitives/Element.js
	var require_Element = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var isEqual = require_isEqual();
		var KeyValuePair = require_KeyValuePair();
		var ArraySlice = require_ArraySlice();
		module.exports = class Element {
			constructor(content, meta, attributes) {
				if (meta) this.meta = meta;
				if (attributes) this.attributes = attributes;
				this.content = content;
			}
			/**
			* Freezes the element to prevent any mutation.
			* A frozen element will add `parent` property to every child element
			* to allow traversing up the element tree.
			*/
			freeze() {
				if (Object.isFrozen(this)) return;
				if (this._meta) {
					this.meta.parent = this;
					this.meta.freeze();
				}
				if (this._attributes) {
					this.attributes.parent = this;
					this.attributes.freeze();
				}
				this.children.forEach((element) => {
					element.parent = this;
					element.freeze();
				}, this);
				if (this.content && Array.isArray(this.content)) Object.freeze(this.content);
				Object.freeze(this);
			}
			primitive() {}
			/**
			* Creates a deep clone of the instance
			*/
			clone() {
				const copy = new this.constructor();
				copy.element = this.element;
				if (this.meta.length) copy._meta = this.meta.clone();
				if (this.attributes.length) copy._attributes = this.attributes.clone();
				if (this.content) if (this.content.clone) copy.content = this.content.clone();
				else if (Array.isArray(this.content)) copy.content = this.content.map((element) => element.clone());
				else copy.content = this.content;
				else copy.content = this.content;
				return copy;
			}
			/**
			*/
			toValue() {
				if (this.content instanceof Element) return this.content.toValue();
				if (this.content instanceof KeyValuePair) return {
					key: this.content.key.toValue(),
					value: this.content.value ? this.content.value.toValue() : void 0
				};
				if (this.content && this.content.map) return this.content.map((element) => element.toValue(), this);
				return this.content;
			}
			/**
			* Creates a reference pointing at the Element
			* @returns {RefElement}
			* @memberof Element.prototype
			*/
			toRef(path) {
				if (this.id.toValue() === "") throw Error("Cannot create reference to an element that does not contain an ID");
				const ref = new this.RefElement(this.id.toValue());
				if (path) ref.path = path;
				return ref;
			}
			/**
			* Finds the given elements in the element tree.
			* When providing multiple element names, you must first freeze the element.
			*
			* @param names {...elementNames}
			* @returns {ArraySlice}
			*/
			findRecursive(...elementNames) {
				if (arguments.length > 1 && !this.isFrozen) throw new Error("Cannot find recursive with multiple element names without first freezing the element. Call `element.freeze()`");
				const elementName = elementNames.pop();
				let elements = new ArraySlice();
				const append = (array, element) => {
					array.push(element);
					return array;
				};
				const checkElement = (array, element) => {
					if (element.element === elementName) array.push(element);
					const items = element.findRecursive(elementName);
					if (items) items.reduce(append, array);
					if (element.content instanceof KeyValuePair) {
						if (element.content.key) checkElement(array, element.content.key);
						if (element.content.value) checkElement(array, element.content.value);
					}
					return array;
				};
				if (this.content) {
					if (this.content.element) checkElement(elements, this.content);
					if (Array.isArray(this.content)) this.content.reduce(checkElement, elements);
				}
				if (!elementNames.isEmpty) elements = elements.filter((element) => {
					let parentElements = element.parents.map((e) => e.element);
					for (const namesIndex in elementNames) {
						const name = elementNames[namesIndex];
						const index = parentElements.indexOf(name);
						if (index !== -1) parentElements = parentElements.splice(0, index);
						else return false;
					}
					return true;
				});
				return elements;
			}
			set(content) {
				this.content = content;
				return this;
			}
			equals(value) {
				return isEqual(this.toValue(), value);
			}
			getMetaProperty(name, value) {
				if (!this.meta.hasKey(name)) {
					if (this.isFrozen) {
						const element = this.refract(value);
						element.freeze();
						return element;
					}
					this.meta.set(name, value);
				}
				return this.meta.get(name);
			}
			setMetaProperty(name, value) {
				this.meta.set(name, value);
			}
			/**
			* @type String
			*/
			get element() {
				return this._storedElement || "element";
			}
			set element(element) {
				this._storedElement = element;
			}
			get content() {
				return this._content;
			}
			set content(value) {
				if (value instanceof Element) this._content = value;
				else if (value instanceof ArraySlice) this.content = value.elements;
				else if (typeof value == "string" || typeof value == "number" || typeof value == "boolean" || value === "null" || value == void 0) this._content = value;
				else if (value instanceof KeyValuePair) this._content = value;
				else if (Array.isArray(value)) this._content = value.map(this.refract);
				else if (typeof value === "object") this._content = Object.keys(value).map((key) => new this.MemberElement(key, value[key]));
				else throw new Error("Cannot set content to given value");
			}
			/**
			* @type ObjectElement
			*/
			get meta() {
				if (!this._meta) {
					if (this.isFrozen) {
						const meta = new this.ObjectElement();
						meta.freeze();
						return meta;
					}
					this._meta = new this.ObjectElement();
				}
				return this._meta;
			}
			set meta(value) {
				if (value instanceof this.ObjectElement) this._meta = value;
				else this.meta.set(value || {});
			}
			/**
			* The attributes property defines attributes about the given instance
			* of the element, as specified by the element property.
			*
			* @type ObjectElement
			*/
			get attributes() {
				if (!this._attributes) {
					if (this.isFrozen) {
						const meta = new this.ObjectElement();
						meta.freeze();
						return meta;
					}
					this._attributes = new this.ObjectElement();
				}
				return this._attributes;
			}
			set attributes(value) {
				if (value instanceof this.ObjectElement) this._attributes = value;
				else this.attributes.set(value || {});
			}
			/**
			* Unique Identifier, MUST be unique throughout an entire element tree.
			* @type StringElement
			*/
			get id() {
				return this.getMetaProperty("id", "");
			}
			set id(element) {
				this.setMetaProperty("id", element);
			}
			/**
			* @type ArrayElement
			*/
			get classes() {
				return this.getMetaProperty("classes", []);
			}
			set classes(element) {
				this.setMetaProperty("classes", element);
			}
			/**
			* Human-readable title of element
			* @type StringElement
			*/
			get title() {
				return this.getMetaProperty("title", "");
			}
			set title(element) {
				this.setMetaProperty("title", element);
			}
			/**
			* Human-readable description of element
			* @type StringElement
			*/
			get description() {
				return this.getMetaProperty("description", "");
			}
			set description(element) {
				this.setMetaProperty("description", element);
			}
			/**
			* @type ArrayElement
			*/
			get links() {
				return this.getMetaProperty("links", []);
			}
			set links(element) {
				this.setMetaProperty("links", element);
			}
			/**
			* Returns whether the element is frozen.
			* @type boolean
			* @see freeze
			*/
			get isFrozen() {
				return Object.isFrozen(this);
			}
			/**
			* Returns all of the parent elements.
			* @type ArraySlice
			*/
			get parents() {
				let { parent } = this;
				const parents = new ArraySlice();
				while (parent) {
					parents.push(parent);
					parent = parent.parent;
				}
				return parents;
			}
			/**
			* Returns all of the children elements found within the element.
			* @type ArraySlice
			* @see recursiveChildren
			*/
			get children() {
				if (Array.isArray(this.content)) return new ArraySlice(this.content);
				if (this.content instanceof KeyValuePair) {
					const children = new ArraySlice([this.content.key]);
					if (this.content.value) children.push(this.content.value);
					return children;
				}
				if (this.content instanceof Element) return new ArraySlice([this.content]);
				return new ArraySlice();
			}
			/**
			* Returns all of the children elements found within the element recursively.
			* @type ArraySlice
			* @see children
			*/
			get recursiveChildren() {
				const children = new ArraySlice();
				this.children.forEach((element) => {
					children.push(element);
					element.recursiveChildren.forEach((child) => {
						children.push(child);
					});
				});
				return children;
			}
		};
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/primitives/NullElement.js
	var require_NullElement = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Element = require_Element();
		/**
		*/
		var NullElement = class extends Element {
			constructor(content, meta, attributes) {
				super(content || null, meta, attributes);
				this.element = "null";
			}
			primitive() {
				return "null";
			}
			set() {
				return /* @__PURE__ */ new Error("Cannot set the value of null");
			}
		};
		module.exports = NullElement;
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/primitives/StringElement.js
	var require_StringElement = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Element = require_Element();
		/**
		* @class StringElement
		*
		* @param {string} content
		* @param meta
		* @param attributes
		*/
		module.exports = class StringElement extends Element {
			constructor(content, meta, attributes) {
				super(content, meta, attributes);
				this.element = "string";
			}
			primitive() {
				return "string";
			}
			/**
			* The length of the string.
			* @type number
			*/
			get length() {
				return this.content.length;
			}
		};
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/primitives/NumberElement.js
	var require_NumberElement = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Element = require_Element();
		/**
		* @class NumberElement
		*
		* @param {number} content
		* @param meta
		* @param attributes
		*/
		module.exports = class NumberElement extends Element {
			constructor(content, meta, attributes) {
				super(content, meta, attributes);
				this.element = "number";
			}
			primitive() {
				return "number";
			}
		};
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/primitives/BooleanElement.js
	var require_BooleanElement = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Element = require_Element();
		/**
		* @class BooleanElement
		*
		* @param {boolean} content
		* @param meta
		* @param attributes
		*/
		module.exports = class BooleanElement extends Element {
			constructor(content, meta, attributes) {
				super(content, meta, attributes);
				this.element = "boolean";
			}
			primitive() {
				return "boolean";
			}
		};
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/primitives/ArrayElement.js
	var require_ArrayElement = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var negate = require_negate();
		var Element = require_Element();
		var ArraySlice = require_ArraySlice();
		/**
		* @class
		*
		* @param {Element[]} content
		* @param meta
		* @param attributes
		*/
		var ArrayElement = class extends Element {
			constructor(content, meta, attributes) {
				super(content || [], meta, attributes);
				this.element = "array";
			}
			primitive() {
				return "array";
			}
			/**
			* @returns {Element}
			*/
			get(index) {
				return this.content[index];
			}
			/**
			* Helper for returning the value of an item
			* This works for both ArrayElement and ObjectElement instances
			*/
			getValue(indexOrKey) {
				const item = this.get(indexOrKey);
				if (item) return item.toValue();
			}
			/**
			* @returns {Element}
			*/
			getIndex(index) {
				return this.content[index];
			}
			set(index, value) {
				this.content[index] = this.refract(value);
				return this;
			}
			remove(index) {
				const removed = this.content.splice(index, 1);
				if (removed.length) return removed[0];
				return null;
			}
			/**
			* @param callback - Function to execute for each element
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			*/
			map(callback, thisArg) {
				return this.content.map(callback, thisArg);
			}
			/**
			* Maps and then flattens the results.
			* @param callback - Function to execute for each element.
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @returns {array}
			*/
			flatMap(callback, thisArg) {
				return this.map(callback, thisArg).reduce((a, b) => a.concat(b), []);
			}
			/**
			* Returns an array containing the truthy results of calling the given transformation with each element of this sequence
			* @param transform - A closure that accepts an element of this array as its argument and returns an optional value.
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @memberof ArrayElement.prototype
			* @returns An array of the non-undefined results of calling transform with each element of the array
			*/
			compactMap(transform, thisArg) {
				const results = [];
				this.forEach((element) => {
					const result = transform.bind(thisArg)(element);
					if (result) results.push(result);
				});
				return results;
			}
			/**
			* @param callback - Function to execute for each element
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @returns {ArraySlice}
			*/
			filter(callback, thisArg) {
				return new ArraySlice(this.content.filter(callback, thisArg));
			}
			/**
			* @param callback - Function to execute for each element
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @returns {ArraySlice}
			*/
			reject(callback, thisArg) {
				return this.filter(negate(callback), thisArg);
			}
			/**
			* This is a reduce function specifically for Minim arrays and objects. It
			* allows for returning normal values or Minim instances, so it converts any
			* primitives on each step.
			*/
			reduce(callback, initialValue) {
				let startIndex;
				let memo;
				if (initialValue !== void 0) {
					startIndex = 0;
					memo = this.refract(initialValue);
				} else {
					startIndex = 1;
					memo = this.primitive() === "object" ? this.first.value : this.first;
				}
				for (let i = startIndex; i < this.length; i += 1) {
					const item = this.content[i];
					if (this.primitive() === "object") memo = this.refract(callback(memo, item.value, item.key, item, this));
					else memo = this.refract(callback(memo, item, i, this));
				}
				return memo;
			}
			/**
			* @callback forEachCallback
			* @param {Element} currentValue
			* @param {NumberElement} index
			*/
			/**
			* @param {forEachCallback} callback - Function to execute for each element
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @memberof ArrayElement.prototype
			*/
			forEach(callback, thisArg) {
				this.content.forEach((item, index) => {
					callback.bind(thisArg)(item, this.refract(index));
				});
			}
			/**
			* @returns {Element}
			*/
			shift() {
				return this.content.shift();
			}
			/**
			* @param value
			*/
			unshift(value) {
				this.content.unshift(this.refract(value));
			}
			/**
			* @param value
			*/
			push(value) {
				this.content.push(this.refract(value));
				return this;
			}
			/**
			* @param value
			*/
			add(value) {
				this.push(value);
			}
			/**
			* Recusively search all descendents using a condition function.
			* @returns {Element[]}
			*/
			findElements(condition, givenOptions) {
				const options = givenOptions || {};
				const recursive = !!options.recursive;
				const results = options.results === void 0 ? [] : options.results;
				this.forEach((item, keyOrIndex, member) => {
					if (recursive && item.findElements !== void 0) item.findElements(condition, {
						results,
						recursive
					});
					if (condition(item, keyOrIndex, member)) results.push(item);
				});
				return results;
			}
			/**
			* Recusively search all descendents using a condition function.
			* @param condition
			* @returns {ArraySlice}
			*/
			find(condition) {
				return new ArraySlice(this.findElements(condition, { recursive: true }));
			}
			/**
			* @param {string} element
			* @returns {ArraySlice}
			*/
			findByElement(element) {
				return this.find((item) => item.element === element);
			}
			/**
			* @param {string} className
			* @returns {ArraySlice}
			* @memberof ArrayElement.prototype
			*/
			findByClass(className) {
				return this.find((item) => item.classes.includes(className));
			}
			/**
			* Search the tree recursively and find the element with the matching ID
			* @param {string} id
			* @returns {Element}
			* @memberof ArrayElement.prototype
			*/
			getById(id) {
				return this.find((item) => item.id.toValue() === id).first;
			}
			/**
			* Looks for matching children using deep equality
			* @param value
			* @returns {boolean}
			*/
			includes(value) {
				return this.content.some((element) => element.equals(value));
			}
			/**
			* Looks for matching children using deep equality
			* @param value
			* @returns {boolean}
			* @see includes
			* @deprecated method was replaced by includes
			*/
			contains(value) {
				return this.includes(value);
			}
			/**
			* @returns {ArrayElement} An empty array element
			*/
			empty() {
				return new this.constructor([]);
			}
			["fantasy-land/empty"]() {
				return this.empty();
			}
			/**
			* @param {ArrayElement} other
			* @returns {ArrayElement}
			*/
			concat(other) {
				return new this.constructor(this.content.concat(other.content));
			}
			["fantasy-land/concat"](other) {
				return this.concat(other);
			}
			["fantasy-land/map"](transform) {
				return new this.constructor(this.map(transform));
			}
			["fantasy-land/chain"](transform) {
				return this.map((element) => transform(element), this).reduce((a, b) => a.concat(b), this.empty());
			}
			["fantasy-land/filter"](callback) {
				return new this.constructor(this.content.filter(callback));
			}
			["fantasy-land/reduce"](transform, initialValue) {
				return this.content.reduce(transform, initialValue);
			}
			/**
			* Returns the length of the collection
			* @type number
			*/
			get length() {
				return this.content.length;
			}
			/**
			* Returns whether the collection is empty
			* @type boolean
			*/
			get isEmpty() {
				return this.content.length === 0;
			}
			/**
			* Return the first item in the collection
			* @type Element
			*/
			get first() {
				return this.getIndex(0);
			}
			/**
			* Return the second item in the collection
			* @type Element
			*/
			get second() {
				return this.getIndex(1);
			}
			/**
			* Return the last item in the collection
			* @type Element
			*/
			get last() {
				return this.getIndex(this.length - 1);
			}
		};
		/**
		* @returns {ArrayElement} An empty array element
		*/
		ArrayElement.empty = function empty() {
			return new this();
		};
		ArrayElement["fantasy-land/empty"] = ArrayElement.empty;
		if (typeof Symbol !== "undefined") ArrayElement.prototype[Symbol.iterator] = function symbol() {
			return this.content[Symbol.iterator]();
		};
		module.exports = ArrayElement;
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/primitives/MemberElement.js
	var require_MemberElement = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var KeyValuePair = require_KeyValuePair();
		var Element = require_Element();
		/**
		* @class MemberElement
		*
		* @param {Element} key
		* @param {Element} value
		* @param meta
		* @param attributes
		*/
		module.exports = class MemberElement extends Element {
			constructor(key, value, meta, attributes) {
				super(new KeyValuePair(), meta, attributes);
				this.element = "member";
				this.key = key;
				this.value = value;
			}
			/**
			* @type Element
			*/
			get key() {
				return this.content.key;
			}
			set key(key) {
				this.content.key = this.refract(key);
			}
			/**
			* @type Element
			*/
			get value() {
				return this.content.value;
			}
			set value(value) {
				this.content.value = this.refract(value);
			}
		};
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/ObjectSlice.js
	var require_ObjectSlice = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var negate = require_negate();
		var ArraySlice = require_ArraySlice();
		module.exports = class ObjectSlice extends ArraySlice {
			map(callback, thisArg) {
				return this.elements.map((member) => callback.bind(thisArg)(member.value, member.key, member));
			}
			filter(callback, thisArg) {
				return new ObjectSlice(this.elements.filter((member) => callback.bind(thisArg)(member.value, member.key, member)));
			}
			reject(callback, thisArg) {
				return this.filter(negate(callback.bind(thisArg)));
			}
			forEach(callback, thisArg) {
				return this.elements.forEach((member, index) => {
					callback.bind(thisArg)(member.value, member.key, member, index);
				});
			}
			/**
			* @returns {array}
			*/
			keys() {
				return this.map((value, key) => key.toValue());
			}
			/**
			* @returns {array}
			*/
			values() {
				return this.map((value) => value.toValue());
			}
		};
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/primitives/ObjectElement.js
	var require_ObjectElement = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var negate = require_negate();
		var isObject = require_isObject();
		var ArrayElement = require_ArrayElement();
		var MemberElement = require_MemberElement();
		var ObjectSlice = require_ObjectSlice();
		/**
		* @class
		*
		* @param content
		* @param meta
		* @param attributes
		*/
		var ObjectElement = class extends ArrayElement {
			constructor(content, meta, attributes) {
				super(content || [], meta, attributes);
				this.element = "object";
			}
			primitive() {
				return "object";
			}
			toValue() {
				return this.content.reduce((results, el) => {
					results[el.key.toValue()] = el.value ? el.value.toValue() : void 0;
					return results;
				}, {});
			}
			/**
			* @param key
			* @returns {Element}
			*/
			get(name) {
				const member = this.getMember(name);
				if (member) return member.value;
			}
			/**
			* @param key
			* @returns {MemberElement}
			*/
			getMember(name) {
				if (name === void 0) return;
				return this.content.find((element) => element.key.toValue() === name);
			}
			/**
			* @param key
			*/
			remove(name) {
				let removed = null;
				this.content = this.content.filter((item) => {
					if (item.key.toValue() === name) {
						removed = item;
						return false;
					}
					return true;
				});
				return removed;
			}
			/**
			* @param key
			* @returns {Element}
			*/
			getKey(name) {
				const member = this.getMember(name);
				if (member) return member.key;
			}
			/**
			* Set allows either a key/value pair to be given or an object
			* If an object is given, each key is set to its respective value
			*/
			set(keyOrObject, value) {
				if (isObject(keyOrObject)) {
					Object.keys(keyOrObject).forEach((objectKey) => {
						this.set(objectKey, keyOrObject[objectKey]);
					});
					return this;
				}
				const key = keyOrObject;
				const member = this.getMember(key);
				if (member) member.value = value;
				else this.content.push(new MemberElement(key, value));
				return this;
			}
			/**
			*/
			keys() {
				return this.content.map((item) => item.key.toValue());
			}
			/**
			*/
			values() {
				return this.content.map((item) => item.value.toValue());
			}
			/**
			* @returns {boolean}
			*/
			hasKey(value) {
				return this.content.some((member) => member.key.equals(value));
			}
			/**
			* @returns {array}
			*/
			items() {
				return this.content.map((item) => [item.key.toValue(), item.value.toValue()]);
			}
			/**
			* @param callback
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			*/
			map(callback, thisArg) {
				return this.content.map((item) => callback.bind(thisArg)(item.value, item.key, item));
			}
			/**
			* Returns an array containing the truthy results of calling the given transformation with each element of this sequence
			* @param transform - A closure that accepts the value, key and member element of this object as its argument and returns an optional value.
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			* @returns An array of the non-undefined results of calling transform with each element of the array
			*/
			compactMap(callback, thisArg) {
				const results = [];
				this.forEach((value, key, member) => {
					const result = callback.bind(thisArg)(value, key, member);
					if (result) results.push(result);
				});
				return results;
			}
			/**
			* @param callback
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			*
			* @returns {ObjectSlice}
			*/
			filter(callback, thisArg) {
				return new ObjectSlice(this.content).filter(callback, thisArg);
			}
			/**
			* @param callback
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			*
			* @returns {ObjectSlice}
			*
			* @memberof ObjectElement.prototype
			*/
			reject(callback, thisArg) {
				return this.filter(negate(callback), thisArg);
			}
			/**
			* @param callback
			* @param thisArg - Value to use as this (i.e the reference Object) when executing callback
			*
			* @memberof ObjectElement.prototype
			*/
			forEach(callback, thisArg) {
				return this.content.forEach((item) => callback.bind(thisArg)(item.value, item.key, item));
			}
		};
		module.exports = ObjectElement;
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/elements/LinkElement.js
	var require_LinkElement = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Element = require_Element();
		/** Hyperlinking MAY be used to link to other resources, provide links to
		* instructions on how to process a given element (by way of a profile or
		* other means), and may be used to provide meta data about the element in
		* which it's found. The meaning and purpose of the hyperlink is defined by
		* the link relation according to RFC 5988.
		*
		* @class LinkElement
		*
		* @param content
		* @param meta
		* @param attributes
		*/
		module.exports = class LinkElement extends Element {
			constructor(content, meta, attributes) {
				super(content || [], meta, attributes);
				this.element = "link";
			}
			/**
			* The relation identifier for the link, as defined in RFC 5988.
			* @type StringElement
			*/
			get relation() {
				return this.attributes.get("relation");
			}
			set relation(relation) {
				this.attributes.set("relation", relation);
			}
			/**
			* The URI for the given link.
			* @type StringElement
			*/
			get href() {
				return this.attributes.get("href");
			}
			set href(href) {
				this.attributes.set("href", href);
			}
		};
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/elements/RefElement.js
	var require_RefElement = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Element = require_Element();
		/**
		* @class RefElement
		*
		* @param content
		* @param meta
		* @param attributes
		*
		* @extends Element
		*/
		module.exports = class RefElement extends Element {
			constructor(content, meta, attributes) {
				super(content || [], meta, attributes);
				this.element = "ref";
				if (!this.path) this.path = "element";
			}
			/**
			* Path of referenced element to transclude instead of element itself.
			* @type StringElement
			* @default element
			*/
			get path() {
				return this.attributes.get("path");
			}
			set path(newValue) {
				this.attributes.set("path", newValue);
			}
		};
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/elements.js
	var require_elements = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var Element = require_Element();
		var NullElement = require_NullElement();
		var StringElement = require_StringElement();
		var NumberElement = require_NumberElement();
		var BooleanElement = require_BooleanElement();
		var ArrayElement = require_ArrayElement();
		var MemberElement = require_MemberElement();
		var ObjectElement = require_ObjectElement();
		var LinkElement = require_LinkElement();
		var RefElement = require_RefElement();
		var ArraySlice = require_ArraySlice();
		var ObjectSlice = require_ObjectSlice();
		var KeyValuePair = require_KeyValuePair();
		/**
		* Refracts a JSON type to minim elements
		* @param value
		* @returns {Element}
		*/
		function refract(value) {
			if (value instanceof Element) return value;
			if (typeof value === "string") return new StringElement(value);
			if (typeof value === "number") return new NumberElement(value);
			if (typeof value === "boolean") return new BooleanElement(value);
			if (value === null) return new NullElement();
			if (Array.isArray(value)) return new ArrayElement(value.map(refract));
			if (typeof value === "object") return new ObjectElement(value);
			return value;
		}
		Element.prototype.ObjectElement = ObjectElement;
		Element.prototype.RefElement = RefElement;
		Element.prototype.MemberElement = MemberElement;
		Element.prototype.refract = refract;
		ArraySlice.prototype.refract = refract;
		/**
		* Contains all of the element classes, and related structures and methods
		* for handling with element instances.
		*/
		module.exports = {
			Element,
			NullElement,
			StringElement,
			NumberElement,
			BooleanElement,
			ArrayElement,
			MemberElement,
			ObjectElement,
			LinkElement,
			RefElement,
			refract,
			ArraySlice,
			ObjectSlice,
			KeyValuePair
		};
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/Namespace.js
	var require_Namespace = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var isNull = require_isNull();
		var isString = require_isString();
		var isNumber = require_isNumber();
		var isBoolean = require_isBoolean();
		var isObject = require_isObject();
		var JSONSerialiser = require_JSONSerialiser();
		var elements = require_elements();
		/**
		* @class
		*
		* A refract element implementation with an extensible namespace, able to
		* load other namespaces into it.
		*
		* The namespace allows you to register your own classes to be instantiated
		* when a particular refract element is encountered, and allows you to specify
		* which elements get instantiated for existing Javascript objects.
		*/
		var Namespace = class {
			constructor(options) {
				this.elementMap = {};
				this.elementDetection = [];
				this.Element = elements.Element;
				this.KeyValuePair = elements.KeyValuePair;
				if (!options || !options.noDefault) this.useDefault();
				this._attributeElementKeys = [];
				this._attributeElementArrayKeys = [];
			}
			/**
			* Use a namespace plugin or load a generic plugin.
			*
			* @param plugin
			*/
			use(plugin) {
				if (plugin.namespace) plugin.namespace({ base: this });
				if (plugin.load) plugin.load({ base: this });
				return this;
			}
			useDefault() {
				this.register("null", elements.NullElement).register("string", elements.StringElement).register("number", elements.NumberElement).register("boolean", elements.BooleanElement).register("array", elements.ArrayElement).register("object", elements.ObjectElement).register("member", elements.MemberElement).register("ref", elements.RefElement).register("link", elements.LinkElement);
				this.detect(isNull, elements.NullElement, false).detect(isString, elements.StringElement, false).detect(isNumber, elements.NumberElement, false).detect(isBoolean, elements.BooleanElement, false).detect(Array.isArray, elements.ArrayElement, false).detect(isObject, elements.ObjectElement, false);
				return this;
			}
			/**
			* Register a new element class for an element.
			*
			* @param {string} name
			* @param elementClass
			*/
			register(name, ElementClass) {
				this._elements = void 0;
				this.elementMap[name] = ElementClass;
				return this;
			}
			/**
			* Unregister a previously registered class for an element.
			*
			* @param {string} name
			*/
			unregister(name) {
				this._elements = void 0;
				delete this.elementMap[name];
				return this;
			}
			detect(test, ElementClass, givenPrepend) {
				if (givenPrepend === void 0 ? true : givenPrepend) this.elementDetection.unshift([test, ElementClass]);
				else this.elementDetection.push([test, ElementClass]);
				return this;
			}
			toElement(value) {
				if (value instanceof this.Element) return value;
				let element;
				for (let i = 0; i < this.elementDetection.length; i += 1) {
					const test = this.elementDetection[i][0];
					const ElementClass = this.elementDetection[i][1];
					if (test(value)) {
						element = new ElementClass(value);
						break;
					}
				}
				return element;
			}
			getElementClass(element) {
				const ElementClass = this.elementMap[element];
				if (ElementClass === void 0) return this.Element;
				return ElementClass;
			}
			fromRefract(doc) {
				return this.serialiser.deserialise(doc);
			}
			toRefract(element) {
				return this.serialiser.serialise(element);
			}
			get elements() {
				if (this._elements === void 0) {
					this._elements = { Element: this.Element };
					Object.keys(this.elementMap).forEach((name) => {
						const pascal = name[0].toUpperCase() + name.substr(1);
						this._elements[pascal] = this.elementMap[name];
					});
				}
				return this._elements;
			}
			/**
			* Convinience method for getting a JSON Serialiser configured with the
			* current namespace
			*
			* @type JSONSerialiser
			* @readonly
			*
			* @memberof Namespace.prototype
			*/
			get serialiser() {
				return new JSONSerialiser(this);
			}
		};
		JSONSerialiser.prototype.Namespace = Namespace;
		module.exports = Namespace;
	}));
	//#endregion
	//#region ../../node_modules/minim/lib/serialisers/JSON06Serialiser.js
	var require_JSON06Serialiser = /* @__PURE__ */ __commonJSMin(((exports, module) => {
		var JSONSerialiser = require_JSONSerialiser();
		module.exports = class JSON06Serialiser extends JSONSerialiser {
			serialise(element) {
				if (!(element instanceof this.namespace.elements.Element)) throw new TypeError(`Given element \`${element}\` is not an Element instance`);
				let variable;
				if (element._attributes && element.attributes.get("variable")) variable = element.attributes.get("variable");
				const payload = { element: element.element };
				if (element._meta && element._meta.length > 0) payload.meta = this.serialiseObject(element.meta);
				const isEnum = element.element === "enum" || element.attributes.keys().indexOf("enumerations") !== -1;
				if (isEnum) {
					const attributes = this.enumSerialiseAttributes(element);
					if (attributes) payload.attributes = attributes;
				} else if (element._attributes && element._attributes.length > 0) {
					let { attributes } = element;
					if (attributes.get("metadata")) {
						attributes = attributes.clone();
						attributes.set("meta", attributes.get("metadata"));
						attributes.remove("metadata");
					}
					if (element.element === "member" && variable) {
						attributes = attributes.clone();
						attributes.remove("variable");
					}
					if (attributes.length > 0) payload.attributes = this.serialiseObject(attributes);
				}
				if (isEnum) payload.content = this.enumSerialiseContent(element, payload);
				else if (this[`${element.element}SerialiseContent`]) payload.content = this[`${element.element}SerialiseContent`](element, payload);
				else if (element.content !== void 0) {
					let content;
					if (variable && element.content.key) {
						content = element.content.clone();
						content.key.attributes.set("variable", variable);
						content = this.serialiseContent(content);
					} else content = this.serialiseContent(element.content);
					if (this.shouldSerialiseContent(element, content)) payload.content = content;
				} else if (this.shouldSerialiseContent(element, element.content) && element instanceof this.namespace.elements.Array) payload.content = [];
				return payload;
			}
			shouldSerialiseContent(element, content) {
				if (element.element === "parseResult" || element.element === "httpRequest" || element.element === "httpResponse" || element.element === "category" || element.element === "link") return true;
				if (content === void 0) return false;
				if (Array.isArray(content) && content.length === 0) return false;
				return true;
			}
			refSerialiseContent(element, payload) {
				delete payload.attributes;
				return {
					href: element.toValue(),
					path: element.path.toValue()
				};
			}
			sourceMapSerialiseContent(element) {
				return element.toValue();
			}
			dataStructureSerialiseContent(element) {
				return [this.serialiseContent(element.content)];
			}
			enumSerialiseAttributes(element) {
				const attributes = element.attributes.clone();
				const enumerations = attributes.remove("enumerations") || new this.namespace.elements.Array([]);
				const defaultValue = attributes.get("default");
				let samples = attributes.get("samples") || new this.namespace.elements.Array([]);
				if (defaultValue && defaultValue.content) {
					if (defaultValue.content.attributes) defaultValue.content.attributes.remove("typeAttributes");
					attributes.set("default", new this.namespace.elements.Array([defaultValue.content]));
				}
				samples.forEach((sample) => {
					if (sample.content && sample.content.element) sample.content.attributes.remove("typeAttributes");
				});
				if (element.content && enumerations.length !== 0) samples.unshift(element.content);
				samples = samples.map((sample) => {
					if (sample instanceof this.namespace.elements.Array) return [sample];
					return new this.namespace.elements.Array([sample.content]);
				});
				if (samples.length) attributes.set("samples", samples);
				if (attributes.length > 0) return this.serialiseObject(attributes);
			}
			enumSerialiseContent(element) {
				if (element._attributes) {
					const enumerations = element.attributes.get("enumerations");
					if (enumerations && enumerations.length > 0) return enumerations.content.map((enumeration) => {
						const e = enumeration.clone();
						e.attributes.remove("typeAttributes");
						return this.serialise(e);
					});
				}
				if (element.content) {
					const value = element.content.clone();
					value.attributes.remove("typeAttributes");
					return [this.serialise(value)];
				}
				return [];
			}
			deserialise(value) {
				if (typeof value === "string") return new this.namespace.elements.String(value);
				if (typeof value === "number") return new this.namespace.elements.Number(value);
				if (typeof value === "boolean") return new this.namespace.elements.Boolean(value);
				if (value === null) return new this.namespace.elements.Null();
				if (Array.isArray(value)) return new this.namespace.elements.Array(value.map(this.deserialise, this));
				const ElementClass = this.namespace.getElementClass(value.element);
				const element = new ElementClass();
				if (element.element !== value.element) element.element = value.element;
				if (value.meta) this.deserialiseObject(value.meta, element.meta);
				if (value.attributes) this.deserialiseObject(value.attributes, element.attributes);
				const content = this.deserialiseContent(value.content);
				if (content !== void 0 || element.content === null) element.content = content;
				if (element.element === "enum") {
					if (element.content) element.attributes.set("enumerations", element.content);
					let samples = element.attributes.get("samples");
					element.attributes.remove("samples");
					if (samples) {
						const existingSamples = samples;
						samples = new this.namespace.elements.Array();
						existingSamples.forEach((existingSample) => {
							existingSample.forEach((sample) => {
								const enumElement = new ElementClass(sample);
								enumElement.element = element.element;
								samples.push(enumElement);
							});
						});
						const sample = samples.shift();
						if (sample) element.content = sample.content;
						else element.content = void 0;
						element.attributes.set("samples", samples);
					} else element.content = void 0;
					let defaultValue = element.attributes.get("default");
					if (defaultValue && defaultValue.length > 0) {
						defaultValue = defaultValue.get(0);
						const defaultElement = new ElementClass(defaultValue);
						defaultElement.element = element.element;
						element.attributes.set("default", defaultElement);
					}
				} else if (element.element === "dataStructure" && Array.isArray(element.content)) [element.content] = element.content;
				else if (element.element === "category") {
					const metadata = element.attributes.get("meta");
					if (metadata) {
						element.attributes.set("metadata", metadata);
						element.attributes.remove("meta");
					}
				} else if (element.element === "member" && element.key && element.key._attributes && element.key._attributes.getValue("variable")) {
					element.attributes.set("variable", element.key.attributes.get("variable"));
					element.key.attributes.remove("variable");
				}
				return element;
			}
			serialiseContent(content) {
				if (content instanceof this.namespace.elements.Element) return this.serialise(content);
				if (content instanceof this.namespace.KeyValuePair) {
					const pair = { key: this.serialise(content.key) };
					if (content.value) pair.value = this.serialise(content.value);
					return pair;
				}
				if (content && content.map) return content.map(this.serialise, this);
				return content;
			}
			deserialiseContent(content) {
				if (content) {
					if (content.element) return this.deserialise(content);
					if (content.key) {
						const pair = new this.namespace.KeyValuePair(this.deserialise(content.key));
						if (content.value) pair.value = this.deserialise(content.value);
						return pair;
					}
					if (content.map) return content.map(this.deserialise, this);
				}
				return content;
			}
			shouldRefract(element) {
				if (element._attributes && element.attributes.keys().length || element._meta && element.meta.keys().length) return true;
				if (element.element === "enum") return false;
				if (element.element !== element.primitive() || element.element === "member") return true;
				return false;
			}
			convertKeyToRefract(key, item) {
				if (this.shouldRefract(item)) return this.serialise(item);
				if (item.element === "enum") return this.serialiseEnum(item);
				if (item.element === "array") return item.map((subItem) => {
					if (this.shouldRefract(subItem) || key === "default") return this.serialise(subItem);
					if (subItem.element === "array" || subItem.element === "object" || subItem.element === "enum") return subItem.children.map((subSubItem) => this.serialise(subSubItem));
					return subItem.toValue();
				});
				if (item.element === "object") return (item.content || []).map(this.serialise, this);
				return item.toValue();
			}
			serialiseEnum(element) {
				return element.children.map((item) => this.serialise(item));
			}
			serialiseObject(obj) {
				const result = {};
				obj.forEach((value, key) => {
					if (value) {
						const keyValue = key.toValue();
						result[keyValue] = this.convertKeyToRefract(keyValue, value);
					}
				});
				return result;
			}
			deserialiseObject(from, to) {
				Object.keys(from).forEach((key) => {
					to.set(key, this.deserialise(from[key]));
				});
			}
		};
	}));
	//#endregion
	//#region ../apidom-core/src/elements/Annotation.mjs
	var import_minim = (/* @__PURE__ */ __commonJSMin(((exports) => {
		require_Namespace();
		var elements = require_elements();
		exports.KeyValuePair = require_KeyValuePair();
		exports.ArraySlice = elements.ArraySlice;
		exports.ObjectSlice = elements.ObjectSlice;
		exports.Element = elements.Element;
		exports.StringElement = elements.StringElement;
		exports.NumberElement = elements.NumberElement;
		exports.BooleanElement = elements.BooleanElement;
		exports.NullElement = elements.NullElement;
		exports.ArrayElement = elements.ArrayElement;
		exports.ObjectElement = elements.ObjectElement;
		exports.MemberElement = elements.MemberElement;
		exports.RefElement = elements.RefElement;
		exports.LinkElement = elements.LinkElement;
		exports.refract = elements.refract;
		exports.JSONSerialiser = require_JSONSerialiser();
		exports.JSON06Serialiser = require_JSON06Serialiser();
	})))();
	/**
	* @public
	*/
	var Annotation = class extends import_minim.StringElement {
		constructor(content, meta, attributes) {
			super(content, meta, attributes);
			this.element = "annotation";
		}
		get code() {
			return this.attributes.get("code");
		}
		set code(value) {
			this.attributes.set("code", value);
		}
	};
	//#endregion
	//#region ../apidom-core/src/elements/Comment.mjs
	/**
	* @public
	*/
	var Comment = class extends import_minim.StringElement {
		constructor(content, meta, attributes) {
			super(content, meta, attributes);
			this.element = "comment";
		}
	};
	//#endregion
	//#region ../apidom-core/src/elements/ParseResult.mjs
	/**
	* @public
	*/
	var ParseResult = class extends import_minim.ArrayElement {
		constructor(content, meta, attributes) {
			super(content, meta, attributes);
			this.element = "parseResult";
		}
		get api() {
			return this.children.filter((item) => item.classes.contains("api")).first;
		}
		get results() {
			return this.children.filter((item) => item.classes.contains("result"));
		}
		get result() {
			return this.results.first;
		}
		get annotations() {
			return this.children.filter((item) => item.element === "annotation");
		}
		get warnings() {
			return this.children.filter((item) => item.element === "annotation" && item.classes.contains("warning"));
		}
		get errors() {
			return this.children.filter((item) => item.element === "annotation" && item.classes.contains("error"));
		}
		get isEmpty() {
			return this.children.reject((item) => item.element === "annotation").isEmpty;
		}
		replaceResult(replacement) {
			const { result } = this;
			if (isUndefined(result)) return false;
			const searchIndex = this.content.findIndex((e) => e === result);
			if (searchIndex === -1) return false;
			this.content[searchIndex] = replacement;
			return true;
		}
	};
	//#endregion
	//#region ../apidom-core/src/predicates/helpers.mjs
	/**
	* @public
	*/
	/**
	* @public
	*/
	/**
	* @public
	*/
	/**
	* @public
	*/
	/**
	* @public
	*/
	/**
	* @public
	*/
	/**
	* @public
	*/
	/**
	* @public
	*/
	var hasMethod = (name, element) => {
		return typeof element === "object" && element !== null && name in element && typeof element[name] === "function";
	};
	/**
	* @public
	*/
	var hasBasicElementProps = (element) => typeof element === "object" && element != null && "_storedElement" in element && typeof element._storedElement === "string" && "_content" in element;
	/**
	* @public
	*/
	var primitiveEq = (val, element) => {
		if (typeof element === "object" && element !== null && "primitive" in element) return typeof element.primitive === "function" && element.primitive() === val;
		return false;
	};
	/**
	* @public
	*/
	var hasClass = (cls, element) => {
		return typeof element === "object" && element !== null && "classes" in element && (Array.isArray(element.classes) || element.classes instanceof import_minim.ArrayElement) && element.classes.includes(cls);
	};
	/**
	* @public
	*/
	var isElementType = (name, element) => typeof element === "object" && element !== null && "element" in element && element.element === name;
	/**
	* @public
	*/
	var createPredicate = (predicateCreator) => {
		return predicateCreator({
			hasMethod,
			hasBasicElementProps,
			primitiveEq,
			isElementType,
			hasClass
		});
	};
	createPredicate(({ hasBasicElementProps, primitiveEq }) => {
		return (element) => element instanceof import_minim.Element || hasBasicElementProps(element) && primitiveEq(void 0, element);
	});
	createPredicate(({ hasBasicElementProps, primitiveEq }) => {
		return (element) => element instanceof import_minim.StringElement || hasBasicElementProps(element) && primitiveEq("string", element);
	});
	createPredicate(({ hasBasicElementProps, primitiveEq }) => {
		return (element) => element instanceof import_minim.NumberElement || hasBasicElementProps(element) && primitiveEq("number", element);
	});
	createPredicate(({ hasBasicElementProps, primitiveEq }) => {
		return (element) => element instanceof import_minim.NullElement || hasBasicElementProps(element) && primitiveEq("null", element);
	});
	createPredicate(({ hasBasicElementProps, primitiveEq }) => {
		return (element) => element instanceof import_minim.BooleanElement || hasBasicElementProps(element) && primitiveEq("boolean", element);
	});
	/**
	* @public
	*/
	var isObjectElement = createPredicate(({ hasBasicElementProps, primitiveEq, hasMethod }) => {
		return (element) => element instanceof import_minim.ObjectElement || hasBasicElementProps(element) && primitiveEq("object", element) && hasMethod("keys", element) && hasMethod("values", element) && hasMethod("items", element);
	});
	/**
	* @public
	*/
	var isArrayElement = createPredicate(({ hasBasicElementProps, primitiveEq, hasMethod }) => {
		return (element) => element instanceof import_minim.ArrayElement && !(element instanceof import_minim.ObjectElement) || hasBasicElementProps(element) && primitiveEq("array", element) && hasMethod("push", element) && hasMethod("unshift", element) && hasMethod("map", element) && hasMethod("reduce", element);
	});
	createPredicate(({ hasBasicElementProps, isElementType, primitiveEq }) => {
		return (element) => element instanceof import_minim.MemberElement || hasBasicElementProps(element) && isElementType("member", element) && primitiveEq(void 0, element);
	});
	createPredicate(({ hasBasicElementProps, isElementType, primitiveEq }) => {
		return (element) => element instanceof import_minim.LinkElement || hasBasicElementProps(element) && isElementType("link", element) && primitiveEq(void 0, element);
	});
	createPredicate(({ hasBasicElementProps, isElementType, primitiveEq }) => {
		return (element) => element instanceof import_minim.RefElement || hasBasicElementProps(element) && isElementType("ref", element) && primitiveEq(void 0, element);
	});
	createPredicate(({ hasBasicElementProps, isElementType, primitiveEq }) => {
		return (element) => element instanceof Annotation || hasBasicElementProps(element) && isElementType("annotation", element) && primitiveEq("array", element);
	});
	createPredicate(({ hasBasicElementProps, isElementType, primitiveEq }) => {
		return (element) => element instanceof Comment || hasBasicElementProps(element) && isElementType("comment", element) && primitiveEq("string", element);
	});
	createPredicate(({ hasBasicElementProps, isElementType, primitiveEq }) => {
		return (element) => element instanceof ParseResult || hasBasicElementProps(element) && isElementType("parseResult", element) && primitiveEq("array", element);
	});
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/evaluate/realms/apidom/realm.mjs
	var ApiDOMEvaluationRealm = class extends EvaluationRealm {
		name = "apidom";
		isArray(node) {
			return isArrayElement(node);
		}
		isObject(node) {
			return isObjectElement(node);
		}
		sizeOf(node) {
			if (this.isArray(node) || this.isObject(node)) return node.length;
			return 0;
		}
		has(node, referenceToken) {
			if (this.isArray(node)) {
				const index = Number(referenceToken);
				const indexUint32 = index >>> 0;
				if (index !== indexUint32) throw new JSONPointerIndexError(`Invalid array index "${referenceToken}": index must be an unsinged 32-bit integer`, {
					referenceToken,
					currentValue: node,
					realm: this.name
				});
				return indexUint32 < this.sizeOf(node);
			}
			if (this.isObject(node)) {
				const keys = node.keys();
				const uniqueKeys = new Set(keys);
				if (keys.length !== uniqueKeys.size) throw new JSONPointerKeyError(`Object key "${referenceToken}" is not unique — JSON Pointer requires unique member names`, {
					referenceToken,
					currentValue: node,
					realm: this.name
				});
				return node.hasKey(referenceToken);
			}
			return false;
		}
		evaluate(node, referenceToken) {
			if (this.isArray(node)) return node.get(Number(referenceToken));
			return node.get(referenceToken);
		}
	};
	//#endregion
	//#region ../../node_modules/@swaggerexpert/json-pointer/es/evaluate/realms/apidom/evaluate.mjs
	var evaluate = (value, jsonPointer, options = {}) => {
		return evaluate$1(value, jsonPointer, {
			...options,
			realm: new ApiDOMEvaluationRealm()
		});
	};
	//#endregion
	exports.ASTTranslator = ASTTranslator;
	exports.CSTTranslator = CSTTranslator;
	exports.Grammar = grammar;
	exports.JSONPointerCompileError = JSONPointerCompileError;
	exports.JSONPointerError = JSONPointerError;
	exports.JSONPointerEvaluateError = JSONPointerEvaluateError;
	exports.JSONPointerIndexError = JSONPointerIndexError;
	exports.JSONPointerKeyError = JSONPointerKeyError;
	exports.JSONPointerParseError = JSONPointerParseError;
	exports.JSONPointerTypeError = JSONPointerTypeError;
	Object.defineProperty(exports, "JSONString", {
		enumerable: true,
		get: function() {
			return json_string_exports;
		}
	});
	Object.defineProperty(exports, "URIFragmentIdentifier", {
		enumerable: true,
		get: function() {
			return uri_fragment_identifier_exports;
		}
	});
	exports.XMLTranslator = XMLTranslator;
	exports.compile = compile;
	exports.escape = escape;
	exports.evaluate = evaluate;
	exports.parse = parse;
	exports.testArrayDash = testArrayDash;
	exports.testArrayIndex = testArrayIndex;
	exports.testArrayLocation = testArrayLocation;
	exports.testJSONPointer = testJSONPointer;
	exports.testReferenceToken = testReferenceToken;
	exports.unescape = unescape;
});
