// @swagger-api/apidom-error/dist/apidom-error.browser.js
(function(global, factory) {
	typeof exports === "object" && typeof module !== "undefined" ? factory(exports) : typeof define === "function" && define.amd ? define(["exports"], factory) : (global = typeof globalThis !== "undefined" ? globalThis : global || self, factory(global.apidomError = {}));
})(this, function(exports) {
	Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
	//#region src/ApiDOMAggregateError.ts
	/**
	* @public
	*/
	var ApiDOMAggregateError = class extends AggregateError {
		constructor(errors, message, options) {
			super(errors, message, options);
			this.name = this.constructor.name;
			if (typeof message === "string") this.message = message;
			if (typeof Error.captureStackTrace === "function") Error.captureStackTrace(this, this.constructor);
			else this.stack = new Error(message).stack;
			/**
			* This needs to stay here until our minimum supported version of Node.js is >= 16.9.0.
			* Node.js >= 16.9.0 supports error causes natively.
			*/
			if (options != null && typeof options === "object" && Object.hasOwn(options, "cause") && !("cause" in this)) {
				const { cause } = options;
				this.cause = cause;
				if (cause instanceof Error && "stack" in cause) this.stack = `${this.stack}\nCAUSE: ${cause.stack}`;
			}
		}
	};
	//#endregion
	//#region src/ApiDOMError.ts
	/**
	* @public
	*/
	var ApiDOMError = class extends Error {
		static [Symbol.hasInstance](instance) {
			return super[Symbol.hasInstance](instance) || Function.prototype[Symbol.hasInstance].call(ApiDOMAggregateError, instance);
		}
		constructor(message, options) {
			super(message, options);
			this.name = this.constructor.name;
			if (typeof message === "string") this.message = message;
			if (typeof Error.captureStackTrace === "function") Error.captureStackTrace(this, this.constructor);
			else this.stack = new Error(message).stack;
			/**
			* This needs to stay here until our minimum supported version of Node.js is >= 16.9.0.
			* Node.js is >= 16.9.0 supports error causes natively.
			*/
			if (options != null && typeof options === "object" && Object.hasOwn(options, "cause") && !("cause" in this)) {
				const { cause } = options;
				this.cause = cause;
				if (cause instanceof Error && "stack" in cause) this.stack = `${this.stack}\nCAUSE: ${cause.stack}`;
			}
		}
	};
	//#endregion
	//#region src/ApiDOMStructuredError.ts
	/**
	* @public
	*/
	var ApiDOMStructuredError = class extends ApiDOMError {
		constructor(message, structuredOptions) {
			super(message, structuredOptions);
			if (structuredOptions != null && typeof structuredOptions === "object") {
				const { cause, ...causelessOptions } = structuredOptions;
				Object.assign(this, causelessOptions);
			}
		}
	};
	//#endregion
	//#region src/UnsupportedOperationError.ts
	/**
	* @public
	*/
	var UnsupportedOperationError = class extends ApiDOMError {};
	//#endregion
	//#region src/NotImplementedError.ts
	/**
	* @public
	*/
	var NotImplementedError = class extends UnsupportedOperationError {};
	//#endregion
	exports.ApiDOMAggregateError = ApiDOMAggregateError;
	exports.ApiDOMError = ApiDOMError;
	exports.ApiDOMStructuredError = ApiDOMStructuredError;
	exports.NotImplementedError = NotImplementedError;
	exports.UnsupportedOperationError = UnsupportedOperationError;
});
